<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Quality control for GWAS summary statistics with minimal filtering • tidyGWAS</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Quality control for GWAS summary statistics with minimal filtering">
<meta name="description" content="tidyGWAS employs a version of dbSNP transformed into Apache Arrow parquet files to restore and validate genome wide association summary statistics.">
<meta property="og:description" content="tidyGWAS employs a version of dbSNP transformed into Apache Arrow parquet files to restore and validate genome wide association summary statistics.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">tidyGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.9.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/tidyGWAS.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/transforming_dbsnp_to_parquet.html">get-to-know-dbsnp</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="tidygwas">tidyGWAS<a class="anchor" aria-label="anchor" href="#tidygwas"></a>
</h1></div>
<!-- badges: start -->

<p>Interested in the trying tidyGWAS out? Check out the get <a href="https://ararder.github.io/tidyGWAS/articles/tidyGWAS.html" class="external-link">started page</a>.</p>
<p>Genome-wide summary statistics are becoming a staple in many different genetics and genomics analysis pipelines. Often, the specific filters suggested for pipelines can be different, requiring each pipeline to have a step where summary statistics are “munged”.</p>
<p>tidyGWAS aims to provide a standardized format <em>before</em> any pipeline specific munging is done. With that in mind, tidyGWAS is conservative in removing rows, and by default keeps both indels and multi-allelic variants.</p>
<p><code>tidyGWAS</code> does the following:</p>
<ol style="list-style-type: decimal">
<li><p>Detection of duplicated rows (based on RSID_REF_ALT or CHR_POS_REF_ALT)</p></li>
<li><p>Standardized column names</p></li>
<li><p>Automatic updating of <a href="https://www.ncbi.nlm.nih.gov/books/NBK573473/" class="external-link">merged</a> RSIDs</p></li>
<li><p>Detection and optional removal of deletions/insertions (“indels”)</p></li>
<li><p>Detection of non rsID values in RSID column, and automatic parsing of the common CHR:POS or CHR:POS:REF:ALT format</p></li>
<li><p>Standardization of CHR values (ex: “23” -&gt; “X”, “chr1” -&gt; “1”)</p></li>
<li>
<p>Validation of standard GWAS columns, B, SE, P, N, FREQ, Z, CaseN, ControlN, A1, A2</p>
<ol style="list-style-type: decimal">
<li>Extremely small pvalues are by default converted to 2.225074e-308 (minimum pvalue in R)</li>
</ol>
</li>
<li><p>Imputation of missing columns: RSID from CHR:POS or CHR:POS from RSID. Any of B,SE, P, Z, N and EAF if missing and possible</p></li>
<li><p>Validation of CHR:POS:RSID by matching with dbSNP v.155</p></li>
<li><p>Cleaned sumstats are provided with coordinates on both GRCh37 and GRCh38, with TRUE/FALSE flags for indels and variants that are multi-allelic in the dataset</p></li>
</ol>
<p>From working with standardized GWAS formats, we’ve found that having both GRCh37 and GRCh38 coordinates, and standardized column names significantly speeds up downstream analysis.</p>
<p>The computationally intensive part of aligning summary statistics with dbSNP 155 (&gt; 940 million rows) for both GRCh37 and GRCh38 (in total 1.8 billion rows) is implemented using the <a href="https://arrow.apache.org/docs/r/" class="external-link">Apache Arrow R</a> implementation, allowing for the full function to run in &lt;3 minutes, using less than 16gb, with ~7 million rows on a Macbook Pro M2.</p>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing tidyGWAS</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Arvid Harder <br><small class="roles"> Maintainer </small>   </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://app.codecov.io/gh/Ararder/tidyGWAS?branch=main" class="external-link"><img src="https://codecov.io/gh/Ararder/tidyGWAS/branch/main/graph/badge.svg" alt="Codecov test coverage"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Arvid Harder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
