---
title: "tidyGWAS"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{tidyGWAS}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(tidyGWAS)
```

# Installation

tidyGWAS contains some wrapper functions around BSgenome and the [SNPlocs](https://bioconductor.org/packages/release/data/annotation/html/SNPlocs.Hsapiens.dbSNP155.GRCh38.html) bioconductor packages. These are quite large dependencies, almost \~14GB, and are therefore listed as "suggests".


# Quick start

```{r, eval=FALSE}

devtools::install_github("ararder/tidyGWAS")
# or
remotes::install_github("ararder/tidyGWAS")


# if you want to be able to repair CHR:POS or RSID, you need to download the following
# bioconductor packages:
if (!require("BiocManager", quietly = TRUE)) install.packages("BiocManager")
BiocManager::install(version = "3.16")
BiocManager::install("BSgenome")
BiocManager::install("Biostrings")
BiocManager::install("GenomicRanges")
BiocManager::install("BSgenome.Hsapiens.1000genomes.hs37d5")
BiocManager::install("BSgenome.Hsapiens.NCBI.GRCh38")
BiocManager::install("SNPlocs.Hsapiens.dbSNP155.GRCh37")
BiocManager::install("SNPlocs.Hsapiens.dbSNP155.GRCh38")


```

```{r, eval = FALSE}
# download RsMergeArch
file <- googledrive::as_dribble("https://drive.google.com/file/d/1HaES_q8m6rSDDLXvYr0TjQ-ZlyhdZbFX/view?usp=share_link")
googledrive::drive_download(file)
```

Put the file wherever you want in your enviroment, but specify the absolute filepath
in .Renviron file, using "
```{r, eval = FALSE}
# usethis::edit_r_environ()
# set "rs_merge_arch"=YOUR_FILE_PATH
# restart R and check that this works:
data.table::fread(Sys.getenv("rs_merge_arch"), nrows = 100)
```

```{r, eval = FALSE}
bs <- get_bsgenome()
rs_merge_arch <- data.table::fread(Sys.getenv("rs_merge_arch"))
tidyGWAS(tbl = my_sumstats, rs_merge_arch = rs_merge_arch,  bsgenome_objects = bs, name = "first_tidyGWAS")


```

# Getting started

Automatic parsing of column names has not been implemented (and probable never will), so you will have to rename the input dataframe to the tidyGWAS format. The following column names are accepted.

```{r}

snp_cols <- c("CHR", "POS", "RSID", "EffectAllele", "OtherAllele", "rowid")
info_cols <- c("INFO", "N", "CaseN", "ControlN", "EAF")
stats_cols <- c("B", "Z", "OR", "P", "SE")
```

How the example file looks. The mandatory columns are either RSID, or CHR and POS

```{r}
test_file <- tidyGWAS::test_file
test_file |>
  dplyr::tibble()
```

The [BSgenome](https://bioconductor.org/packages/release/bioc/html/BSgenome.html) and [SNPlocs](https://bioconductor.org/packages/release/data/annotation/html/SNPlocs.Hsapiens.dbSNP155.GRCh38.html) packages are used to store and query dbSNP and fasta files to get CHR, POS, RSID, REF and ALT for all SNPs in dbSNP v155. The first query to this dataframe will always be slower than subsequent queries, so whenever you want to munge several summary statistics, it's get to first load the points to the bsgenome and snplocs data, which can be done with `get_bsgenome().`

```{r,eval = FALSE}
bsgenome <- get_bsgenome()

```

You can skip the step to align with dbSNP, which will significantly speed up running time and reduce memory usage, by not providing `bsgnome_objects`. To update the RSID colum, you also need to provide an RsMergeArch file. tidyGWAS will attempt to read in this file using a filepath provided in the R enviroment "rs_merge_arch".

```{r, eval=FALSE}
# You can download the rs_merge_arch file here:
file <- googledrive::as_dribble("https://drive.google.com/file/d/1HaES_q8m6rSDDLXvYr0TjQ-ZlyhdZbFX/view?usp=share_link")
googledrive::drive_download(file)
# use usethis to set the env variable rs_merge_arch to the filepath to the downloaded file 
# usethis::edit_r_environ()

```

```{r}
# use the small example file stored with tidyGWAS package
rs_merge_arch <- tidyGWAS::rs_merge_arch

```

```{r, collapse=TRUE}
start <- Sys.time()
tmp <- tidyGWAS(test_file, rs_merge_arch = rs_merge_arch)
end <- Sys.time()
```

About 3 seconds for 100k rows

```{r, collapse=TRUE}
end - start


```

If you want the information to be stored in a logfile on disk, you can use the logfile flag. Note that by default, tidyGWAS uses a folder in `tempdir` to store all intermediate files. These files are not saved indefinitely, so to store the output data, you can also provide an output directory where you want the copy over the finished tidyGWAS oflder

```{r, echo = FALSE}
my_outdir <- withr::local_tempdir("all_my_gwas_files")
tidyGWAS(test_file, rs_merge_arch = rs_merge_arch, logfile = TRUE, outdir = my_outdir)

```

If you dont provide a name for the sumstat cleaning, tidyGWAS uses `Sys.time()` to create a random name.

```{r}
list.files(my_outdir, recursive = TRUE)

```

This can be controlled with the `name` argument:

```{r, echo = FALSE}
tidyGWAS(test_file, rs_merge_arch = rs_merge_arch, logfile = TRUE, outdir = my_outdir, name = "first_gwas")
list.files(my_outdir, recursive = TRUE)

```

Oops! We now have two different folders in our outdir. This is a feature. `outdir` is meant to be used when you have many different sumstats that you all want to be in the same directory, with the same folder structure.

## The functions within

tidyGWAS exposes several smaller functions that can help when interactively working with sumstats.

## Getting B and SE from Z, N and EAF

`repair_stats()` can be used to add any missing statistics column, such as P, Z, B or SE, depending on which columns exist. Note that there some

```{r}
# Has Z , EAF and N, but not B and SE (Which are commonly used for PRS)
tmp <- dplyr::mutate(test_file, Z = B/SE, N = CaseN + ControlN) |> 
  dplyr::select(-B, -SE)
repaired <- repair_stats(tmp)

# the Betas will be on a standardized scale, and can therefore be different
# than what was initially there
dplyr::inner_join(test_file, repaired,by = "RSID") |> dplyr::select(B.x, B.y) |> 
  dplyr::summarise(cor(B.x, B.y))


# Getting Z, if you only have B and P

```

```{r}
tmp <-  dplyr::select(test_file,  -SE) |> 
  repair_stats()
```

## Identifying incorrect RSIDs

uses a regex to identify any non RrSs rows.

```{r, eval = FALSE}
flag_incorrect_rsid_format(test_file) |> 
  dplyr::filter(invalid_rsid)

```

## A fast function for finding duplications in RSID or CHR

```{r}
flag_duplicates(test_file, column = "rsid")
# or 
flag_duplicates(test_file, column = "chr_pos")


```

### working with dbSNP
Sometimes the same CHR and POS maps to multiple RSIDs.
Currently, we resolve this by slecting the smallest RSID
```{r}
#  CHR        POS RSID         ref_allele alt_alleles
#   <chr>    <int> <chr>        <chr>      <list>     
# 1 1     39491595 rs10157617   T          <chr [2]>  
# 2 1     39491595 rs1638449573 T          <chr [1]>  
# 3 1     39491595 rs1638449625 T          <chr [1]>  
# 4 1     39491595 rs1638449683 T          <chr [1]>  

```

