---
title: "tidyGWAS"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{tidyGWAS}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Quick start

### Installation

```{r, eval = FALSE}
devtools::install_github("ararder/tidyGWAS") # install.packages("devtools")
remotes::install_github("ararder/tidyGWAS") # install.packages("remotes")
```

### Download reference files - This only has to be done once

```{r, eval = FALSE}
library(tidyGWAS)
library(fs)

# Provide filepath to a directory where you want to store reference files
ref_dir <- "path/to/storage/directory"
download_ref_files(ref_dir)

# the reference files will be stored in ref_dir/dbSNP155
dbsnp_files <- paste0(ref_dir, "/dbSNP155")
```

### Provide column names tidyGWAS understands

tidyGWAS does NOT automatically parse common column names, so you need
to provide it with the correct column names.

If your RSID column is in the format CHR:BP:A1:A2, you can still pass it
as RSID.

If N is missing, please add it to the input file prior to running
tidyGWAS.

the tidyGWAS column names:

-   CHR

-   POS

-   RSID

-   EffectAllele

-   OtherAllele

-   EAF

-   B

-   SE

-   P

-   CaseN

-   ControlN

-   N

-   INFO

-   Z

It's very common to have a wide variety of column names used in summary
statistics. `tidyGWAS()` uses the `column_map` argument to correctly
parse arguments. Pass a named lists, with the names corresponding to the
tidyGWAS column and the value to the column name in the file.

You can pass either a filepath or a in-memory data.frame as the first
argument.

```{r}
# creating a example summary statistics file with other column names used than tidyGWAS
library(tidyGWAS)
library(fs)

sumstats <- tidyGWAS::test_file

wrong_names <- dplyr::rename(sumstats, chrom = CHR, A1 = EffectAllele, AlleleFreq = EAF)
head(wrong_names) |> 
  knitr::kable()
```

tidyGWAS would not be able to correctly parse these column names. In
this specific case, tidyGWAS throws and error because EffectAllele and
OtherAllele are required to proceed.

```{r, eval=FALSE}
tidyGWAS(
  wrong_names,
  # small test set of dbSNP reference data
  fs::path(fs::path_package("tidyGWAS"), "extdata/dbSNP155")
  )
```

To fix this, we provide the `column_map` argument with a named list. The
list names should correspond to tidyGWAS column names, and the values to
the column names in the input file.

```{r}
colmap <- list(
  "CHR" = "chrom",
  "EffectAllele" = "A1",
  "EAF" = "AlleleFreq"
)

cleaned <- tidyGWAS(
  tbl = wrong_names, 
  # provide the filepath to the refence files you downloaded previously
  # and add dbSNP155
  dbsnp_path = fs::path(fs::path_package("tidyGWAS"), "extdata/dbSNP155"),
  column_map = colmap
  )


```

It's not uncommon for sample size to also be missing from the summary
statistics. These can be passed in a similar manner as the column names.
Here we use the `sample_size_map` to create new sample size columns

```{r}
colmap <- list(
  "CHR" = "chrom",
  "EffectAllele" = "A1",
  "EAF" = "AlleleFreq"
)
smap <- list(
  CaseN = 34000,
  ControlN = 73000
)

cleaned <- tidyGWAS(
  tbl = wrong_names, 
  dbsnp_path = fs::path(fs::path_package("tidyGWAS"), "extdata/dbSNP155"),
  column_map = colmap,
  # now we add the sample_size_map argument
  sample_size_map = smap
  )

dplyr::select(cleaned, CaseN, ControlN, N) |> 
  head() |> 
  knitr::kable()

```

# In depth walk-through

Let's walk through each step more carefully, using some example files
provided in the tidyGWAS package. We will be using a version of the
sumstats from the latest GWAS on schizophrenia from the Psychiatric
Genomics Consortium
[(PGC3)](https://www.nature.com/articles/s41586-022-04434-5)

```{r}
library(tidyGWAS)
library(fs)

example_sumstats <- tidyGWAS::test_file
# here we use the example files provided in the tidyGWAS package
example_ref_file <-  fs::path(fs::path_package("tidyGWAS"), "extdata/dbSNP155")

#The columns have been renamed to follow the tidyGWAS format:

colnames(example_sumstats)
```

The two mandatory argument is `tbl` and `dbsnp_path`. tidyGWAS accepts
either a in-memory `base::data.frame()` or a filepath to text file. If
you pass a filepath to a text file, tidyGWAS will attempt to read it
using `arrow::read_delim_arrow()`.

It's not uncommon for sumstats to have formats that is not supported by
this function, in which case i recommend you read in the file using
`data.table::fread()` or `readr::read_tsv()` first, and then pass the
`data.frame` to `tidyGWAS()`.

### A more typical workflow

Typically, you will call `tidyGWAS()` with a few more arguments than the
example above. By default `tidyGWAS()` will use `base::tempdir()` to
write to during execution. In addition, the cleaned sumstat is also
returned by the function. This means that files will NOT be saved after
you quit the R session. This behavior can be adjusted with the `outdir`
argument, where files will be saved in `outdir` after a successful run.

Let's take a look at how a more typical call to tidyGWAS would look:

1.  `logfile` can be set to TRUE, so the output of tidyGWAS will be
    saved in a logfile
2.  `name` The output directory can be named as a convenience.
3.  `output_format` tidyGWAS supports three output formats: 'csv',
    'parquet' or 'hivestyle'
4.  `outdir` a filepath to a directory where you want to store the
    output. This folder should not exist before running tidyGWAS. If it
    does, tidyGWAS will throw an error to prevent overwriting of files

## example

Usually, the columnm names will not correspond to tidyGWAS column names.
For this case, lets also pretend that there was no sample size columns,
and that the sumstats are stored with tab-separation in a gzip file.

```{r}
# Passing a filepath also works
tmp_path <- fs::path_temp("example_sumstats.tsv.gz")
example_sumstats <- example_sumstats |> 
  dplyr::rename(
  CHROM = CHR, BP = POS, A1 = EffectAllele, Effect=B
) |> 
dplyr::select(-CaseN, -ControlN)

readr::write_tsv(example_sumstats, tmp_path)
```

```{r}
# we setup a directory where we will store our tidyGWAS output
gwas_folder <- tempfile()

cleaned <- tidyGWAS(
  # provide the filepath to my summary statistics
  tbl = tmp_path, 
  # we need to provide the file delimiter
  delim = "\t",
  # filepath to the reference files you downloaded previously
  dbsnp_path = example_ref_file,
  # we input the column names of the input file that do not correspond to tidyGWAS
  column_map = list(
    "CHR" = "CHROM",
    "POS" = "BP",
    "EffectAllele" = "A1",
    "B" = "Effect"
  ),
  # we also pass in a list with the number of cases and controls since it's missing.
  sample_size_map = list(
    CaseN = 54000,
    ControlN = 73000
  ),
  # it's useful to write all the messages to a logfile, so messages are saved.
  logfile=TRUE,
  # i want the output to be in the parquet format
  output_format = "parquet",
  # i want the output folder to be named after which GWAS it is. 
  outdir = paste0(gwas_folder, "/scz_pgc3"),
  )


```

If we take a look in the output folder we provided, we can see that a
new folder has been created, we can see the standard files that tidyGWAS
will output.

1.  `tidyGWAS_cleaned.parquet` contains the cleaned summary statistics

2.  `scz_pgc3/metadata.yaml` contains meta data, such that any run can
    be reproduced.

    ```{r}
    yaml::read_yaml(fs::path(gwas_folder, "scz_pgc3/metadata.yaml"))
    ```

3.  `scz_pgc3/raw/example_sumstats.tsv.gz.parquet` contains the raw file
    before processing

4.  `scz_pgc3/tidyGWAS_logfile.txt` is the log file.

```{r}
list.files(gwas_folder, recursive = TRUE)
```

```{r}

```

# Parallel computation

tidGWAS automatically detects the number of cores, and uses it to
parallelize the alignment with dbSNP. In some cases, for example when
running tidyGWAS in a HPC cluster, you might need to manually set the
number of cores, which can be done using the `OMP_NUM_THREADS` variable.
This should not be a larger number of cores than what you have requested
in your HPC job (in the example below, the "--cpus-per-task" flag)

```{bash, eval=FALSE}

#SBATCH --mem=60gb
#SBATCH --time=24:0:00
#SBATCH --cpus-per-task 8
export OMP_NUM_THREADS=8

outdir=$(pwd)
gwas=$outdir/my_gwas.tsv.gz
dbsnp_files="dbSNP155"
Rscript -e "tidyGWAS(commandArgs(trailingOnly = TRUE)[1],  dbsnp_path = commandArgs(trailingOnly = TRUE)[2],outdir = commandArgs(trailingOnly = TRUE)[3], logfile=TRUE)" $gwas $dbsnp_files $outdir

```

# Computational cost

Memory use and time scales with the size of the summary statistics. From
running tidyGWAS experimentally, here's an estimation:

The memory usage will likely be the main constraint with increasing size
of summary statistics.

1.  20 million rows \~20gb
2.  40 million rows \~40gb
3.  60 million rows \~75gb

# download_ref_data() is not working

tidyGWAS uses a version dbSNP 155 converted to the [Apache
Arrow](https://arrow.apache.org) .parquet files. You can download the
dbSNP155 reference file from inside R using the `googledrive` package,
or by manually navigating to this
[file](https://drive.google.com/file/d/1LmgfmTQaWwJaFpBHcRQIY_kwe5iN7Pj6/view?usp=share_link)

```{r, eval = FALSE}
# You can download the file from inside R using the googledrive package:
library(googledrive)
googledrive::drive_deauth()
id <- googledrive::as_id("1LmgfmTQaWwJaFpBHcRQIY_kwe5iN7Pj6")

##### EDIT THIS:
filepath_to_store_dir <- ""
##### ---------------------

drive_download(id, path = filepath_to_store_dir)
```

The file needs to be untarred

```{bash, eval = FALSE}
# change directory to where the downloaded file is
tar -xf dbSNP155.tar dbSNP155

```
