<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="`tidyGWAS()` performs a set of validations on input colummns, repairs missing
columns, and can add missing CHR/POS or RSID. In addition, CHR and POS is
standardised to GRCh38, with coordinates on GRCh37 added in as well.
Briefly, `tidyGWAS()` updates RSID if possible using the
[refsnp-merged](https://ftp.ncbi.nih.gov/snp/latest_release/JSON/) file
from dbSNP. Each inputed column is then validated and coerced to
the correct type. Missing CHR/POS or RSID is detected and imputed using
[repair_rsid()] or [repair_chr_pos()]. If both RSID and CHR:POS is present,
[verify_chr_pos_rsid()] is executed to check that dbSNP CHR:POS:RSID agrees
with CHR:POS:RSID in `tbl`.
If statistis such as `P`, `B` are missing, `tidyGWAS()` will attempt to impute
them if possible using [repair_stats()]
Standard column names are assumed, BEFORE inputting into the function. This is
a deliberate decision, as automatic parsing of some important column names
can be ambigious For example, in some sumstats, A1 referes to effect allele,
while other formats use A1 as non-effect allele. [tidyGWAS_columns()] can be
used to standardise column names, and see the standard format."><title>Execute validation and quality control of GWAS summmary statistics — tidyGWAS • tidyGWAS</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Execute validation and quality control of GWAS summmary statistics — tidyGWAS"><meta property="og:description" content="`tidyGWAS()` performs a set of validations on input colummns, repairs missing
columns, and can add missing CHR/POS or RSID. In addition, CHR and POS is
standardised to GRCh38, with coordinates on GRCh37 added in as well.
Briefly, `tidyGWAS()` updates RSID if possible using the
[refsnp-merged](https://ftp.ncbi.nih.gov/snp/latest_release/JSON/) file
from dbSNP. Each inputed column is then validated and coerced to
the correct type. Missing CHR/POS or RSID is detected and imputed using
[repair_rsid()] or [repair_chr_pos()]. If both RSID and CHR:POS is present,
[verify_chr_pos_rsid()] is executed to check that dbSNP CHR:POS:RSID agrees
with CHR:POS:RSID in `tbl`.
If statistis such as `P`, `B` are missing, `tidyGWAS()` will attempt to impute
them if possible using [repair_stats()]
Standard column names are assumed, BEFORE inputting into the function. This is
a deliberate decision, as automatic parsing of some important column names
can be ambigious For example, in some sumstats, A1 referes to effect allele,
while other formats use A1 as non-effect allele. [tidyGWAS_columns()] can be
used to standardise column names, and see the standard format."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tidyGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/tidyGWAS.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/get-to-know-dbsnp.html">get-to-know-dbsnp</a>
    <a class="dropdown-item" href="../articles/tidyGWAS-workflow.html">tidyGWAS-workflow</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"></ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Execute validation and quality control of GWAS summmary statistics</h1>
      
      <div class="d-none name"><code>tidyGWAS.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>`tidyGWAS()` performs a set of validations on input colummns, repairs missing
columns, and can add missing CHR/POS or RSID. In addition, CHR and POS is
standardised to GRCh38, with coordinates on GRCh37 added in as well.</p>
<p>Briefly, `tidyGWAS()` updates RSID if possible using the
[refsnp-merged](https://ftp.ncbi.nih.gov/snp/latest_release/JSON/) file
from dbSNP. Each inputed column is then validated and coerced to
the correct type. Missing CHR/POS or RSID is detected and imputed using
[repair_rsid()] or [repair_chr_pos()]. If both RSID and CHR:POS is present,
[verify_chr_pos_rsid()] is executed to check that dbSNP CHR:POS:RSID agrees
with CHR:POS:RSID in `tbl`.</p>
<p>If statistis such as `P`, `B` are missing, `tidyGWAS()` will attempt to impute
them if possible using [repair_stats()]</p>
<p>Standard column names are assumed, BEFORE inputting into the function. This is
a deliberate decision, as automatic parsing of some important column names
can be ambigious For example, in some sumstats, A1 referes to effect allele,
while other formats use A1 as non-effect allele. [tidyGWAS_columns()] can be
used to standardise column names, and see the standard format.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">tidyGWAS</span><span class="op">(</span></span>
<span>  <span class="va">tbl</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  <span class="va">dbsnp_path</span>,</span>
<span>  output_format <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"csv"</span>, <span class="st">"hivestyle"</span>, <span class="st">"parquet"</span><span class="op">)</span>,</span>
<span>  build <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NA"</span>, <span class="st">"37"</span>, <span class="st">"38"</span><span class="op">)</span>,</span>
<span>  outdir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  <span class="va">study_n</span>,</span>
<span>  convert_p <span class="op">=</span> <span class="fl">2.225074e-308</span>,</span>
<span>  name <span class="op">=</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace_all</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/date.html" class="external-link">date</a></span><span class="op">(</span><span class="op">)</span>, pattern <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>` ` <span class="op">=</span> <span class="st">"_"</span>, `:` <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  keep_indels <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  repair_cols <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  logfile <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  log_on_err <span class="op">=</span> <span class="st">"tidyGWAS_logfile.txt"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>tbl</dt>
<dd><p>a `data.frame` or `character()` vector</p></dd>


<dt>...</dt>
<dd><p>pass additional arguments to [arrow::read_delim_arrow()] which is the function
that will be called if tbl is a filepath
possible arguments are `study_n` to set N, and `build` to set genome build.</p></dd>


<dt>dbsnp_path</dt>
<dd><p>filepath to the dbSNP155 directory (untarred dbSNP155.tar)</p></dd>


<dt>output_format</dt>
<dd><p>How should the finished cleaned file be saved?
* "csv" corresponds to [arrow::write_csv_arrow()]
* 'hivestyle' corresponds to [arrow::write_dataset()] split by CHR
* 'parquet' corresponds to [arrow::write_parquet()]</p></dd>


<dt>build</dt>
<dd><p>Can be used to skip [infer_build()]</p></dd>


<dt>outdir</dt>
<dd><p>Where should results be saved after a succesful execution?</p></dd>


<dt>study_n</dt>
<dd><p>Use to set N in tbl</p></dd>


<dt>convert_p</dt>
<dd><p>What value should be used for P = 0?</p></dd>


<dt>name</dt>
<dd><p>name of the output directory</p></dd>


<dt>keep_indels</dt>
<dd><p>Should indels be kept?</p></dd>


<dt>repair_cols</dt>
<dd><p>Should any missing columns be repaired?</p></dd>


<dt>logfile</dt>
<dd><p>Should messages be redirected to a logfile?</p></dd>


<dt>log_on_err</dt>
<dd><p>Optional. Can pass a filepath to copy the logfile to when the function exists.</p></dd>


<dt>verbose</dt>
<dd><p>Explain filters in detail?</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>a tibble or NULL, depending on outdir</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="fu">tidyGWAS</span><span class="op">(</span>tbl <span class="op">=</span> <span class="st">"my_dataframe"</span>, logfile <span class="op">=</span> <span class="st">"true"</span>, name <span class="op">=</span> <span class="st">"test_run"</span>, outdir <span class="op">=</span> <span class="st">"gwas_sumstat_dir"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Arvid Harder.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

