<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Execute validation and quality control of GWAS summmary statistics — tidyGWAS • tidyGWAS</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Execute validation and quality control of GWAS summmary statistics — tidyGWAS"><meta name="description" content="tidyGWAS() performs a set of validations on input colummns, repairs missing
columns, and can add missing CHR/POS or RSID. In addition, CHR and POS is
standardised to GRCh38, with coordinates on GRCh37 added in as well.
Briefly, tidyGWAS() updates RSID if possible using the
refsnp-merged file
from dbSNP. Each inputed column is then validated and coerced to
the correct type.
If statistis such as P, B are missing, tidyGWAS() will attempt to impute
them if possible using repair_stats()
Standard column names are assumed, BEFORE inputting into the function. This is
a deliberate decision as automatic parsing of some important column names
can be ambiguous For example, in some sumstats, A1 referes to effect allele,
while other formats use A1 as non-effect allele."><meta property="og:description" content="tidyGWAS() performs a set of validations on input colummns, repairs missing
columns, and can add missing CHR/POS or RSID. In addition, CHR and POS is
standardised to GRCh38, with coordinates on GRCh37 added in as well.
Briefly, tidyGWAS() updates RSID if possible using the
refsnp-merged file
from dbSNP. Each inputed column is then validated and coerced to
the correct type.
If statistis such as P, B are missing, tidyGWAS() will attempt to impute
them if possible using repair_stats()
Standard column names are assumed, BEFORE inputting into the function. This is
a deliberate decision as automatic parsing of some important column names
can be ambiguous For example, in some sumstats, A1 referes to effect allele,
while other formats use A1 as non-effect allele."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidyGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.9.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/tidyGWAS.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/transforming_dbsnp_to_parquet.html">get-to-know-dbsnp</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Execute validation and quality control of GWAS summmary statistics</h1>

      <div class="d-none name"><code>tidyGWAS.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>tidyGWAS()</code> performs a set of validations on input colummns, repairs missing
columns, and can add missing CHR/POS or RSID. In addition, CHR and POS is
standardised to GRCh38, with coordinates on GRCh37 added in as well.</p>
<p>Briefly, <code>tidyGWAS()</code> updates RSID if possible using the
<a href="https://ftp.ncbi.nih.gov/snp/latest_release/JSON/" class="external-link">refsnp-merged</a> file
from dbSNP. Each inputed column is then validated and coerced to
the correct type.</p>
<p>If statistis such as <code>P</code>, <code>B</code> are missing, <code>tidyGWAS()</code> will attempt to impute
them if possible using <code><a href="repair_stats.html">repair_stats()</a></code></p>
<p>Standard column names are assumed, BEFORE inputting into the function. This is
a deliberate decision as automatic parsing of some important column names
can be ambiguous For example, in some sumstats, A1 referes to effect allele,
while other formats use A1 as non-effect allele.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">tidyGWAS</span><span class="op">(</span></span>
<span>  <span class="va">tbl</span>,</span>
<span>  <span class="va">dbsnp_path</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  <span class="va">column_names</span>,</span>
<span>  output_format <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"hivestyle"</span>, <span class="st">"parquet"</span>, <span class="st">"csv"</span><span class="op">)</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  CaseN <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ControlN <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  N <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  build <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"NA"</span>, <span class="st">"37"</span>, <span class="st">"38"</span><span class="op">)</span>,</span>
<span>  default_build <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"37"</span>, <span class="st">"38"</span><span class="op">)</span>,</span>
<span>  indel_strategy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"keep"</span>, <span class="st">"remove"</span><span class="op">)</span>,</span>
<span>  convert_p <span class="op">=</span> <span class="fl">2.225074e-308</span>,</span>
<span>  repair_cols <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  logfile <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-tbl">tbl<a class="anchor" aria-label="anchor" href="#arg-tbl"></a></dt>
<dd><p>a <code>data.frame</code> or <code><a href="https://rdrr.io/r/base/character.html" class="external-link">character()</a></code> vector</p></dd>


<dt id="arg-dbsnp-path">dbsnp_path<a class="anchor" aria-label="anchor" href="#arg-dbsnp-path"></a></dt>
<dd><p>filepath to the dbSNP155 directory (untarred dbSNP155.tar)</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>pass additional arguments to <code><a href="https://arrow.apache.org/docs/r/reference/read_delim_arrow.html" class="external-link">arrow::read_delim_arrow()</a></code>, if tbl is a filepath.</p></dd>


<dt id="arg-column-names">column_names<a class="anchor" aria-label="anchor" href="#arg-column-names"></a></dt>
<dd><p>a named list of column names:
<code>list(RSID = "SNP", POS = "BP")</code></p></dd>


<dt id="arg-output-format">output_format<a class="anchor" aria-label="anchor" href="#arg-output-format"></a></dt>
<dd><p>How should the finished cleaned file be saved?</p><ul><li><p>'csv' corresponds to <code><a href="https://arrow.apache.org/docs/r/reference/write_csv_arrow.html" class="external-link">arrow::write_csv_arrow()</a></code></p></li>
<li><p>'parquet' corresponds to <code><a href="https://arrow.apache.org/docs/r/reference/write_parquet.html" class="external-link">arrow::write_parquet()</a></code></p></li>
<li><p>'hivestyle' corresponds to <code><a href="https://arrow.apache.org/docs/r/reference/write_dataset.html" class="external-link">arrow::write_dataset()</a></code> split by <code>CHR</code></p></li>
</ul></dd>


<dt id="arg-output-dir">output_dir<a class="anchor" aria-label="anchor" href="#arg-output-dir"></a></dt>
<dd><p>filepath to a folder where tidyGWAS output will be stored.
The folder should not yet exist. Note that the default argument is <code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile()</a></code>,
meaning that tidyGWAS output will not be saved by default over R sessions.</p></dd>


<dt id="arg-casen">CaseN<a class="anchor" aria-label="anchor" href="#arg-casen"></a></dt>
<dd><p>manually input number of cases</p></dd>


<dt id="arg-controln">ControlN<a class="anchor" aria-label="anchor" href="#arg-controln"></a></dt>
<dd><p>manually input number of controls</p></dd>


<dt id="arg-n">N<a class="anchor" aria-label="anchor" href="#arg-n"></a></dt>
<dd><p>manually input sample size</p></dd>


<dt id="arg-build">build<a class="anchor" aria-label="anchor" href="#arg-build"></a></dt>
<dd><p>If you are sure of what genome build ('37' or '38'), can be used to skip <code><a href="infer_build.html">infer_build()</a></code> and speed up computation</p></dd>


<dt id="arg-default-build">default_build<a class="anchor" aria-label="anchor" href="#arg-default-build"></a></dt>
<dd><p>If only RSID exists, the build cannot be inferred. Nonetheless,
tidyGWAS applies a filter on incompatible alleles with GRCh37/38. In such a case,
tidyGWAS needs to decide on which reference genome to compare alleles with.</p></dd>


<dt id="arg-indel-strategy">indel_strategy<a class="anchor" aria-label="anchor" href="#arg-indel-strategy"></a></dt>
<dd><p>Should indels be kept or removed?</p></dd>


<dt id="arg-convert-p">convert_p<a class="anchor" aria-label="anchor" href="#arg-convert-p"></a></dt>
<dd><p>What value should be used for when P-value has been rounded to 0?</p></dd>


<dt id="arg-repair-cols">repair_cols<a class="anchor" aria-label="anchor" href="#arg-repair-cols"></a></dt>
<dd><p>Should any missing statistical columns be repaired if possible?</p></dd>


<dt id="arg-logfile">logfile<a class="anchor" aria-label="anchor" href="#arg-logfile"></a></dt>
<dd><p>Should messages be redirected to a logfile?</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a <code><a href="https://dplyr.tidyverse.org/reference/reexports.html" class="external-link">dplyr::tibble()</a></code></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu">tidyGWAS</span><span class="op">(</span>tbl <span class="op">=</span> <span class="st">"my_dataframe"</span>, logfile <span class="op">=</span> <span class="st">"true"</span>, name <span class="op">=</span> <span class="st">"test_run"</span>, outdir <span class="op">=</span> <span class="st">"gwas_sumstat_dir"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Arvid Harder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

