<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>meta-analysis • tidyGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="meta-analysis">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidyGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/tidyGWAS.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/meta-analysis.html">meta-analysis</a></li>
    <li><a class="dropdown-item" href="../articles/transforming_dbsnp_to_parquet.html">get-to-know-dbsnp</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>meta-analysis</h1>
            
      

      <div class="d-none name"><code>meta-analysis.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ararder.github.io/tidyGWAS/">tidyGWAS</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="implementation-details">Implementation details<a class="anchor" aria-label="anchor" href="#implementation-details"></a>
</h2>
<p>tidyGWAS implements functionality to perform inverse-variance
weighted meta-analysis of summary statistics from multiple GWAS, using
[tidyGWAS::meta_analysis()]. The decisions and strategies for tidyGWAS
is described here.</p>
<p><code>tidyGWAS::meta_analysis()</code> requires summary statistics to
first be cleaned with <code><a href="../reference/tidyGWAS.html">tidyGWAS::tidyGWAS()</a></code>, and for the
output_format to be ‘hivestyle’.</p>
<p><code>tidyGWAS::meta_analysis()</code> will first apply a set of
filter across all variants:</p>
<ol style="list-style-type: decimal">
<li>B and SE are required to be real numeric values (not NA, Inf or
NaN)</li>
<li>No missing values in (CHR,RSID, A1,A2)</li>
<li>EffectAllele or OtherAllele has to be the same as the REF*
allele.</li>
</ol>
<p>All alleles are then flipped, such that EffectAllele corresponds to
the REF allele. Effect sizes and allele frequencies are flipped
accordingly.</p>
<p>Variant identity is then defined based on RSID:REF:ALT. This will
handle multi-allelics, but will not handle differences in strand
alignment across summary statistics.</p>
<p>Columns N,CaseN, ControlN, EffectiveN are summed across studies, with
NA values removed.</p>
<p>Columns INFO and EAF are averaged across studies. For a value to be
included in the average, both INFO/EAF and N needs to be present. For
example, if a summary statistics file has INFO, but not N, the final
sample-size weighted will not include that specific study, as the INFO
column could not be weighted by sample size.</p>
<p>tidyGWAS then performs fixed-effects meta-analysis using
inverse-variance weighting with weights calculated based on the standard
error. An optional argument to weigh by sample size is forthcoming.</p>
<p>A few things to keep in mind:</p>
<div class="section level4">
<h4 id="choice-of-reference-allele">Choice of reference allele<a class="anchor" aria-label="anchor" href="#choice-of-reference-allele"></a>
</h4>
<p>The reference allele varies at a locations between GRCh37 and GRCH38.
For example, out of 13,372,985 variants with MAF &gt;= 0.005 in the EUR
subpopulation of 1000G, 64892 variants differ in reference allele
(~0.05% of variants) between builds.</p>
<p>tidyGWAS returns the reference allele from both GRCh37 and GRCh38,
and users can choose which reference allele to use. Since the reference
allele differs, either the alignment of EffectAllele will be different,
or, the variant that is about to be meta_analyzed will be removed, in
the case that neither EffectAllele or OtherAllele is equal to the
reference allele.</p>
</div>
<div class="section level4">
<h4 id="strand-alignment-and-ambigious-snps">Strand alignment and ambigious SNPs<a class="anchor" aria-label="anchor" href="#strand-alignment-and-ambigious-snps"></a>
</h4>
<p>tidyGWAS does not perform any strand alignment beyond flipping the
effect allele to always be the reference allele. If some variants are on
different strands, they will be considered as different variants.</p>
</div>
<div class="section level4">
<h4 id="multi-allelic-variants">Multi-allelic variants<a class="anchor" aria-label="anchor" href="#multi-allelic-variants"></a>
</h4>
<p>Since tidyGWAS uses RSID:REF:ALT to identify variants, it can handle
multi-allelic variants, and the estimation of P and B should be correct.
However, representation of allele frequency (EAF) for multi-allelic
variants has not been standardised in the field. Some formatas will
contain the minor allele frequency only, while other studies choose to
calculate the frequency of the effect allele as (N_A1 / (N_A1 +N_A2)).
When tidyGWAS flips allele frequencies during harmonisation of the
effect allele, this can introduce errors in allele frequencies for
multi-allelic variants.</p>
</div>
</div>
<div class="section level2">
<h2 id="tutorial">Tutorial<a class="anchor" aria-label="anchor" href="#tutorial"></a>
</h2>
<p>tidyGWAS relies on the arrow package to read and</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Arvid Harder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
