<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="tidyGWAS">
<title>tidyGWAS • tidyGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="tidyGWAS">
<meta property="og:description" content="tidyGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tidyGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/tidyGWAS.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/transforming_dbsnp_to_parquet.html">get-to-know-dbsnp</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>tidyGWAS</h1>
            
      
      
      <div class="d-none name"><code>tidyGWAS.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ararder/tidyGWAS"</span><span class="op">)</span> <span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ararder/tidyGWAS"</span><span class="op">)</span> <span class="co"># install.packages("remotes")</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="download-reference-files">Download reference files<a class="anchor" aria-label="anchor" href="#download-reference-files"></a>
</h2>
<p>tidyGWAS uses reference data from dbSNP, transformed into parquet
files. They can be downloaded <a href="https://zenodo.org/records/11060095" class="external-link">here</a>.</p>
<p>The file needs to untarred before it can be used.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># zenodo link</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">wget</span> <span class="at">-O</span> dbSNP155.tar https://zenodo.org/records/11060095/files/dbSNP155.tar</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xvf</span> dbSNP155.tar</span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co"># remove the tarred dir</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="fu">rm</span> dbSNP155.tar</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>tidyGWAS has two mandatory arguments:</p>
<ul>
<li><p><code>tbl</code>: An in-memory <code>data.frame</code> or a file
path to the summary statistics</p></li>
<li><p><code>dbsnp_path</code>: A file path to the dbSNP reference
file</p></li>
</ul>
<p>An argument that is not mandatory but very useful is
<code>output_dir</code>. tidyGWAS outputs all the files from an
execution in a directory, so the filepath provided to
<code>output_dir</code> is the path to where tidyGWAS will create the
output folder. The folder cannot exist yet (to prevent accidental
overwriting of files).</p>
<p>In the example below, a new directory called
<code>scz_testfile</code> will be created inside
<code>/home/cleaned_tidyGWAS</code>. If <code>output_dir</code>is not
set, tidyGWAS will create a folder inside the R tempdir.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># here we use an in-memory data frame that comes with the tidyGWAS package</span></span>
<span><span class="va">sumstats</span> <span class="op">&lt;-</span> <span class="fu">tidyGWAS</span><span class="fu">::</span><span class="va"><a href="../reference/test_file.html">test_file</a></span></span>
<span><span class="va">output_dir</span> <span class="op">&lt;-</span> <span class="st">"path/to/tidyGWAS_files/&lt;trait_name&gt;"</span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  <span class="co"># Here we input the summary statistics as a data.frame already in R memory</span></span>
<span>  tbl <span class="op">=</span> <span class="va">sumstats</span>, </span>
<span>  <span class="co"># provide the filepath to the refence files you downloaded.</span></span>
<span>  dbsnp_path <span class="op">=</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"tidyGWAS"</span><span class="op">)</span>, <span class="st">"extdata/dbSNP155"</span><span class="op">)</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="st">"/home/cleaned_tidyGWAS/scz_testfile"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="columnar-names">Columnar names<a class="anchor" aria-label="anchor" href="#columnar-names"></a>
</h4>
<p>In the vast majority of cases, the column names in the input summary
statistics will be different from the nomenclature that tidyGWAS uses.
For that case we use the <code>column_names</code> argument, which takes
a <code>list</code> as input.</p>
<p>The list names correspond to tidyGWAS nomenclature, and the values to
the column name in the summary statistics file you want to clean.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First we create a list</span></span>
<span><span class="co"># list names = tidyGWAS names</span></span>
<span><span class="co"># list values = input names</span></span>
<span></span>
<span><span class="va">col_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"CHR"</span> <span class="op">=</span> <span class="st">"chrom"</span>,</span>
<span>  <span class="st">"EffectAllele"</span> <span class="op">=</span> <span class="st">"A1"</span>,</span>
<span>  <span class="st">"EAF"</span> <span class="op">=</span> <span class="st">"AlleleFreq"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">sumstats</span> <span class="op">&lt;-</span> <span class="fu">tidyGWAS</span><span class="fu">::</span><span class="va"><a href="../reference/test_file.html">test_file</a></span></span>
<span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">sumstats</span>, </span>
<span>  <span class="co"># provide the filepath to the refence files you downloaded</span></span>
<span>  dbsnp_path <span class="op">=</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"tidyGWAS"</span><span class="op">)</span>, <span class="st">"extdata/dbSNP155"</span><span class="op">)</span>,</span>
<span>  </span>
<span>  column_names <span class="op">=</span> <span class="va">col_names</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="the-tidygwas-nomenclature">The tidyGWAS nomenclature<a class="anchor" aria-label="anchor" href="#the-tidygwas-nomenclature"></a>
</h4>
<p>tidyGWAS uses the following column names:</p>
<ul>
<li><p>CHR</p></li>
<li><p>POS</p></li>
<li><p>RSID</p></li>
<li><p>EffectAllele</p></li>
<li><p>OtherAllele</p></li>
<li><p>EAF</p></li>
<li><p>B</p></li>
<li><p>SE</p></li>
<li><p>P</p></li>
<li><p>CaseN</p></li>
<li><p>ControlN</p></li>
<li><p>N</p></li>
<li><p>INFO</p></li>
<li><p>Z</p></li>
</ul>
<p>Note:</p>
<p>If your RSID column is in the format CHR:BP:A1:A2, you can still pass
it as an RSID column.</p>
</div>
<div class="section level4">
<h4 id="reading-files-directly-from-disk">Reading files directly from disk<a class="anchor" aria-label="anchor" href="#reading-files-directly-from-disk"></a>
</h4>
<p>If you pass a filepath to tidyGWAS, it will attempt to read in the
file with <code><a href="https://arrow.apache.org/docs/r/reference/read_delim_arrow.html" class="external-link">arrow::read_delim_arrow()</a></code></p>
<p>The default delimiter is white space, so to read in comma-separated
files or tab-separated files, you can use <code>...</code> to provide
the <code>delim</code> argument to
<code><a href="https://arrow.apache.org/docs/r/reference/read_delim_arrow.html" class="external-link">arrow::read_delim_arrow()</a></code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org" class="external-link">readr</a></span><span class="op">)</span></span>
<span><span class="va">tmp_filepath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># write a tsv file to disk</span></span>
<span><span class="fu">write_tsv</span><span class="op">(</span><span class="fu">tidyGWAS</span><span class="fu">::</span><span class="va"><a href="../reference/test_file.html">test_file</a></span>, <span class="va">tmp_filepath</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Here we need to provide the delim argument</span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">tmp_filepath</span>,</span>
<span>  <span class="co"># here we specify the delimiter</span></span>
<span>  delim <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>  dbsnp_path <span class="op">=</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"tidyGWAS"</span><span class="op">)</span>, <span class="st">"extdata/dbSNP155"</span><span class="op">)</span>,</span>
<span><span class="op">)</span></span></code></pre></div>
<p>There’s a lot of different field delimiters used in the wild, and
sometimes you can struggle with inputting the correct delimiter. In such
cases, it’s often much more convenient to use the effective
<code>data.table::fread()</code> or <code>readr::read_table()</code> to
first read in the summary statistics into memory before passing it to
tidyGWAS.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># write a tsv file to disk</span></span>
<span><span class="va">sumstats</span> <span class="op">&lt;-</span> <span class="fu">fread</span><span class="op">(</span><span class="st">"path/to/sumstats.gz.vcf"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Here we need to provide the delim argument</span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">tmp_filepath</span>,</span>
<span>  <span class="co"># here we specify the delimiter</span></span>
<span>  delim <span class="op">=</span> <span class="st">"\t"</span>,</span>
<span>  dbsnp_path <span class="op">=</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"tidyGWAS"</span><span class="op">)</span>, <span class="st">"extdata/dbSNP155"</span><span class="op">)</span>,</span>
<span>  <span class="co"># here we specify the output directory, and use scz2022 as name for the output directory</span></span>
<span>  outdir <span class="op">=</span> <span class="st">"output_dir/scz2022"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="inputting-sample-size-columns">Inputting sample size columns<a class="anchor" aria-label="anchor" href="#inputting-sample-size-columns"></a>
</h4>
<p>It’s not uncommon for sample size to also be missing from the summary
statistics. These can be passed in a similar manner as the column names.
tidyGWAS has three arguments that can be used to manually set the sample
size, if it’s missing from the original file:</p>
<ul>
<li><p><code>CaseN</code></p></li>
<li><p><code>ControlN</code></p></li>
<li><p><code>N</code></p></li>
</ul>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colmap</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"CHR"</span> <span class="op">=</span> <span class="st">"chrom"</span>,</span>
<span>  <span class="st">"EffectAllele"</span> <span class="op">=</span> <span class="st">"A1"</span>,</span>
<span>  <span class="st">"EAF"</span> <span class="op">=</span> <span class="st">"AlleleFreq"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">wrong_names</span>, </span>
<span>  dbsnp_path <span class="op">=</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"tidyGWAS"</span><span class="op">)</span>, <span class="st">"extdata/dbSNP155"</span><span class="op">)</span>,</span>
<span>  column_names <span class="op">=</span> <span class="va">colmap</span>,</span>
<span>  <span class="co"># now we add the sample_size arguments</span></span>
<span>  CaseN <span class="op">=</span> <span class="fl">400</span>, <span class="co"># we can set </span></span>
<span>  ControlN <span class="op">=</span> <span class="fl">800</span>,</span>
<span>  N <span class="op">=</span> <span class="fl">1200</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="a-typical-tidygwas-call">A typical tidyGWAS call<a class="anchor" aria-label="anchor" href="#a-typical-tidygwas-call"></a>
</h2>
<p>In most cases, more than the three first arguments will be used.
Let’s walk through each step more carefully, using some example files
provided in the tidyGWAS package. We will be using a version of the
sumstats from the latest GWAS on schizophrenia from the Psychiatric
Genomics Consortium <a href="https://www.nature.com/articles/s41586-022-04434-5" class="external-link">(PGC3)</a></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://arvidharder.com/tidyGWAS/">tidyGWAS</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org" class="external-link">fs</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">example_sumstats</span> <span class="op">&lt;-</span> <span class="fu">tidyGWAS</span><span class="fu">::</span><span class="va"><a href="../reference/test_file.html">test_file</a></span></span>
<span><span class="co"># here we use the example files provided in the tidyGWAS package</span></span>
<span><span class="va">example_ref_file</span> <span class="op">&lt;-</span>  <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"tidyGWAS"</span><span class="op">)</span>, <span class="st">"extdata/dbSNP155"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#The columns have been renamed to follow the tidyGWAS format:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">example_sumstats</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "CHR"          "POS"          "RSID"         "EffectAllele" "OtherAllele" </span></span>
<span><span class="co">#&gt;  [6] "B"            "SE"           "EAF"          "INFO"         "P"           </span></span>
<span><span class="co">#&gt; [11] "CaseN"        "ControlN"</span></span></code></pre></div>
<p>Change up some of the column names to simulate what would typically
happen.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">example_sumstats</span> <span class="op">&lt;-</span> <span class="va">example_sumstats</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span></span>
<span>  CHROM <span class="op">=</span> <span class="va">CHR</span>, BP <span class="op">=</span> <span class="va">POS</span>, A1 <span class="op">=</span> <span class="va">EffectAllele</span>, Effect<span class="op">=</span><span class="va">B</span></span>
<span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">CaseN</span>, <span class="op">-</span><span class="va">ControlN</span><span class="op">)</span></span></code></pre></div>
<p>When we call tidyGWAS, we need to provide it with the correct column
names.</p>
<p>By default, tidyGWAS produces hivestyle parquet files, but if you
prefer the classical ‘csv’ file, you can use <code>output_format</code>
argument.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we setup a directory where we will store our tidyGWAS output</span></span>
<span><span class="va">gwas_folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  <span class="co"># provide the filepath to my summary statistics, (on disk)</span></span>
<span>  tbl <span class="op">=</span> <span class="va">example_sumstats</span>, </span>
<span>  <span class="co"># we need to provide the file delimiter</span></span>
<span>  delim <span class="op">=</span> <span class="st">","</span>,</span>
<span>  <span class="co"># filepath to the reference files you downloaded previously</span></span>
<span>  dbsnp_path <span class="op">=</span> <span class="va">example_ref_file</span>,</span>
<span>  <span class="co"># we input the column names of the input file that do not correspond to tidyGWAS</span></span>
<span>  column_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"CHR"</span> <span class="op">=</span> <span class="st">"CHROM"</span>,</span>
<span>    <span class="st">"POS"</span> <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>    <span class="st">"EffectAllele"</span> <span class="op">=</span> <span class="st">"A1"</span>,</span>
<span>    <span class="st">"B"</span> <span class="op">=</span> <span class="st">"Effect"</span></span>
<span>  <span class="op">)</span>,</span>
<span>  <span class="co"># This is the default argument, but you can change to 'csv' or 'parquet'</span></span>
<span>  <span class="co"># to output a csv file or a .parquet file</span></span>
<span>  output_format <span class="op">=</span> <span class="st">"hivestyle"</span>,</span>
<span>  <span class="co"># sample sizes were missing, so we need to provide it also</span></span>
<span>  CaseN <span class="op">=</span> <span class="fl">54000</span>,</span>
<span>  ControlN <span class="op">=</span> <span class="fl">73000</span>,</span>
<span>  <span class="co"># output_format can be adjusted. </span></span>
<span>  <span class="co"># ouput_format = "csv"</span></span>
<span>  <span class="co"># it's useful to write all the messages to a logfile, so messages are saved.</span></span>
<span>  logfile<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>  <span class="co"># i want the output folder to be named after which GWAS it was. </span></span>
<span>  output_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">gwas_folder</span>, <span class="st">"/scz_pgc3"</span><span class="op">)</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Running </span><span style="color: #0000BB; font-weight: bold;">tidyGWAS 0.8.0</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; Starting at 2024-05-13 07:27:27.5552</span></span>
<span><span class="co">#&gt; with 100000 rows in input data.frame</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Saving output in folder: <span style="color: #0000BB;">/tmp/RtmpVoDdEl/file19ff4b2b4b57/scz_pgc3</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Output is redirected to logfile: <span style="color: #0000BB;">/tmp/RtmpVoDdEl/file19ff4b2b4b57/scz_pgc3/tidyGWAS_logfile.txt</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking that columns follow tidyGWAS format </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> The following columns are used for further steps: <span style="font-style: italic;">CHR</span>, <span style="font-style: italic;">POS</span>, <span style="font-style: italic;">RSID</span>, <span style="font-style: italic;">EffectAllele</span>, <span style="font-style: italic;">OtherAllele</span>, <span style="font-style: italic;">B</span>, <span style="font-style: italic;">SE</span>, <span style="font-style: italic;">EAF</span>, <span style="font-style: italic;">INFO</span>, <span style="font-style: italic;">P</span>, <span style="font-style: italic;">rowid</span>, <span style="font-style: italic;">CaseN</span>, and <span style="font-style: italic;">ControlN</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking for columns with all NA </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no columns with all NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">1) Scanning for rows with NA</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> No rows contained missing values</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">2) Scanning for rows with duplications</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A unique ID is formed by concontenating CHR, POS, EffectAllele, and OtherAllele</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no duplications</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">3) Scanning for indels</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. EffectAllele or OtherAllele, character length &gt; 1: A vs AA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. EffectAllele or OtherAllele coded as 'D', 'I', or 'R'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Detected 0 rows as indels</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 4a) Validating columns </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The median value of B is -0.0012, which seems reasonable</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-style: italic;">All rows passed validation</span> </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 5) Adding RSID based on CHR:POS. Adding dbSNP based QC flags </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Inferring build by matching 10000 rows to GRCh37 and GRCh38</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3437 snps matched GRCh38, 9998 for GRCh37, inferring build to be 37</span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> Removed 11 rows with no dbSNP entry</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">/tmp/RtmpVoDdEl/file19ff4b2b4b57/scz_pgc3/pipeline_info/removed_nodbsnp.parquet</span></span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> Removed 486 rows with missing on either build</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">/tmp/RtmpVoDdEl/file19ff4b2b4b57/scz_pgc3/pipeline_info/removed_missing_on_either_build.parquet</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">6) Repairing missings statistics columns if possible</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Z missing: Calculating Z using the formula:  Z = B / SE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Finished repair_stats:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Added 1 new columns: Z</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Listing final breakdown of removed rows:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; missing_on_either_build: 486</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; nodbsnp: 11</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Finished tidyGWAS</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A total of 497 rows were removed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Total running time: 4.4s</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Saving metadata from analysis to</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">/tmp/RtmpVoDdEl/file19ff4b2b4b57/scz_pgc3/metadata.yaml</span></span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="the-files-produced-by-tidygwas">The files produced by tidyGWAS<a class="anchor" aria-label="anchor" href="#the-files-produced-by-tidygwas"></a>
</h2>
<p>In the folder specified in <code>output_dir</code>, you will have the
following files:</p>
<ol style="list-style-type: decimal">
<li>
<code>metadata.yaml</code> a record of arguments passed. Useful for
reproducing results.</li>
<li>
<code>pipeline_info/*</code> will contain a set of files
corresponding to removed rows.</li>
<li>
<code>raw/*</code> prior to any cleaning, the raw summary statistics
are saved without edits.</li>
<li>
<code>tidyGWAS_hivestyle</code> If you used the default
output_format, this folder will exist and contain the cleaned summary
statistics in a hivestyle partitioned format with .parquet files.</li>
<li>
<code>tidyGWAS_logfile.txt</code> If you used
<code>logfile=TRUE</code>, the log is saved here</li>
</ol>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_tree.html" class="external-link">dir_tree</a></span><span class="op">(</span><span class="va">gwas_folder</span>, recurse <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co">#&gt; /tmp/RtmpVoDdEl/file19ff4b2b4b57</span></span>
<span><span class="co">#&gt; └── scz_pgc3</span></span>
<span><span class="co">#&gt;     ├── metadata.yaml</span></span>
<span><span class="co">#&gt;     ├── pipeline_info</span></span>
<span><span class="co">#&gt;     ├── raw</span></span>
<span><span class="co">#&gt;     ├── tidyGWAS_hivestyle</span></span>
<span><span class="co">#&gt;     └── tidyGWAS_logfile.txt</span></span></code></pre></div>
<div class="section level4">
<h4 id="hivestyle-partitioning-format">Hivestyle partitioning format<a class="anchor" aria-label="anchor" href="#hivestyle-partitioning-format"></a>
</h4>
<p>The default output format is a hivestyle format using .parquet files.
See for example <a href="https://arrow.apache.org/docs/r/articles/dataset.html" class="external-link">here</a>
for some motivation for this. In essence, this format will significantly
speed up other downstream applications such as meta-analysis, LD
querying and other analyses.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_tree.html" class="external-link">dir_tree</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">gwas_folder</span>,<span class="st">"scz_pgc3"</span>, <span class="st">"tidyGWAS_hivestyle"</span><span class="op">)</span>, recurse <span class="op">=</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; /tmp/RtmpVoDdEl/file19ff4b2b4b57/scz_pgc3/tidyGWAS_hivestyle</span></span>
<span><span class="co">#&gt; ├── CHR=1</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=10</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=11</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=12</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=13</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=14</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=15</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=16</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=17</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=18</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=19</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=2</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=20</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=21</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=22</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=3</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=4</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=5</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=6</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=7</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; ├── CHR=8</span></span>
<span><span class="co">#&gt; │   └── part-0.parquet</span></span>
<span><span class="co">#&gt; └── CHR=9</span></span>
<span><span class="co">#&gt;     └── part-0.parquet</span></span>
<span></span>
<span><span class="co"># This is the equivalent of readr::read_tsv() or data.table::fread() for </span></span>
<span><span class="co"># the hivestyle format.</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="the-output-format">The output format<a class="anchor" aria-label="anchor" href="#the-output-format"></a>
</h2>
<p>The partitioned .parquet files can easily be read into memory using
<code>arrow::open_dataset() |&gt; collect()</code>.</p>
<p><code><a href="../reference/tidyGWAS.html">tidyGWAS()</a></code> will add a set of columns, some depending on
what existed in the input columns.</p>
<ol style="list-style-type: decimal">
<li>If possible, statistics columns such as <code>B, P, SE, Z</code>
will be added if missing and if it is possible to impute them.</li>
<li>
<code>POS, CHR</code> correspond to GRCh38 while
<code>POS_37, CHR_37</code> correspond to GRCh37. Both will always
exist, regardless of initial build, unless
<code>add_missing_build = FALSE</code>.</li>
<li>
<code>multi_allelic</code> is a TRUE/FALSE column that flag rows
that were multi-allelic IN the summary statistics NOT whether there are
multiple alleles in dbSNP. (TRUE corresponds to multi_allelic).</li>
<li>
<code>rowid</code> maps each row back to the inputted summary
statistics, so that any row can be mapped back to it’s original
values.</li>
<li>
<code>REF</code> is the reference genome allele <em>on whichever
build the summary statistics where initially on</em>.</li>
<li>
<code>ID</code> is a concatenation of
CHR:POS:EffectAllele:OtherAllele, where EffectAllele will always
correspond to <code>REF</code>. OBS!! B or EAF has NOT been flipped, so
to use the ID correctly you would first need to flip B and EAF.</li>
<li>
<code>indel</code> is a TRUE/FALSE columns that flags insertions or
deletions. TRUE for rows which are indels. Will be missing unlesss
indels are detected.</li>
<li>The remaining columns are the tidyGWAS nomenclature for standard
GWAS columns</li>
</ol>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">arrow</span><span class="fu">::</span><span class="fu"><a href="https://arrow.apache.org/docs/r/reference/open_dataset.html" class="external-link">open_dataset</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">gwas_folder</span>,<span class="st">"scz_pgc3"</span>, <span class="st">"tidyGWAS_hivestyle"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/compute.html" class="external-link">collect</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">arrow</span><span class="fu">::</span><span class="fu"><a href="https://arrow.apache.org/docs/r/reference/open_dataset.html" class="external-link">open_dataset</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="va">gwas_folder</span>,<span class="st">"scz_pgc3"</span>, <span class="st">"tidyGWAS_hivestyle"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="co"># note: when working interactively, remember to select a subset of columns. This will result in much quicker IO.</span></span>
<span>  <span class="co"># dplyr::select(-dplyr::any_of(c("CHR_37", "POS_37","rowid", "multi_allelic", "indel", "REF")))</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/compute.html" class="external-link">collect</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "POS"           "RSID"          "EffectAllele"  "OtherAllele"  </span></span>
<span><span class="co">#&gt;  [5] "EAF"           "Z"             "B"             "SE"           </span></span>
<span><span class="co">#&gt;  [9] "P"             "N"             "CaseN"         "ControlN"     </span></span>
<span><span class="co">#&gt; [13] "INFO"          "CHR_37"        "POS_37"        "rowid"        </span></span>
<span><span class="co">#&gt; [17] "multi_allelic" "REF"           "ID"            "CHR"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="parallel-computation">Parallel computation<a class="anchor" aria-label="anchor" href="#parallel-computation"></a>
</h2>
<p>tidyGWAS automatically detects the number of cores, and uses it to
parallelize the alignment with dbSNP. In some cases, for example when
running tidyGWAS in a HPC cluster, you might need to manually set the
number of cores, which can be done using the
<code>OMP_NUM_THREADS</code> variable. This should not be a larger
number of cores than what you have requested in your HPC job (in the
example below, the “–cpus-per-task” flag)</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="co">#SBATCH --mem=60gb</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="co">#SBATCH --time=24:0:00</span></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task 8</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="bu">export</span> <span class="va">OMP_NUM_THREADS</span><span class="op">=</span>8</span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a></span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a><span class="va">outdir</span><span class="op">=</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a><span class="va">gwas</span><span class="op">=</span><span class="va">$outdir</span>/my_gwas.tsv.gz</span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a><span class="va">dbsnp_files</span><span class="op">=</span><span class="st">"dbSNP155"</span></span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">"tidyGWAS(commandArgs(trailingOnly = TRUE)[1],  dbsnp_path = commandArgs(trailingOnly = TRUE)[2],outdir = commandArgs(trailingOnly = TRUE)[3], logfile=TRUE)"</span> <span class="va">$gwas</span> <span class="va">$dbsnp_files</span> <span class="va">$outdir</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="computational-cost">Computational cost<a class="anchor" aria-label="anchor" href="#computational-cost"></a>
</h2>
<p>Memory use and time scales with the size of the summary statistics.
From running tidyGWAS experimentally, here’s an estimation:</p>
<p>The memory usage will likely be the main constraint with increasing
size of summary statistics.</p>
<ol style="list-style-type: decimal">
<li>20 million rows ~20gb</li>
<li>40 million rows ~40gb</li>
<li>60 million rows ~75gb</li>
</ol>
</div>
<div class="section level2">
<h2 id="section"></h2>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Arvid Harder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
