<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>tidyGWAS • tidyGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="tidyGWAS">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">tidyGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/tidyGWAS.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/meta-analysis.html">meta-analysis</a></li>
    <li><a class="dropdown-item" href="../articles/transforming_dbsnp_to_parquet.html">get-to-know-dbsnp</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>tidyGWAS</h1>
            
      

      <div class="d-none name"><code>tidyGWAS.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>tidyGWAS is not yet on CRAN. To install, you need to use
<code>devtools::install_github()</code> or
<code>remotes::install_github()</code></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ararder/tidyGWAS"</span><span class="op">)</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ararder/tidyGWAS"</span><span class="op">)</span></span></code></pre></div>
<p>Secondly, you will need to download the reference data. tidyGWAS uses
a slightly edited version of dbSNP v155, that is available <a href="https://zenodo.org/records/14697504" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">wget</span> https://zenodo.org/records/16639374/files/dbSNP155.tar</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xvf</span> dbSNP155.tar</span></code></pre></div>
<div class="section level3">
<h3 id="using-an-apptainer-or-docker-container-instead">Using an apptainer or docker container instead<a class="anchor" aria-label="anchor" href="#using-an-apptainer-or-docker-container-instead"></a>
</h3>
<p>You can skip the installation step by using a docker container with
docker or apptainer.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># using apptainer</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="ex">apptainer</span> pull docker://arvhar/tidygwas:latest</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="ex">apptainer</span> shell ~/tidygwas_latest.sif</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co"># using docker </span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="ex">docker</span> run arvhar/tidygwas:latest</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="news">NEWS<a class="anchor" aria-label="anchor" href="#news"></a>
</h2>
<p>With the release of <strong>tidyGWAS 1.0</strong>, the package now
supports automatic column-name guessing as well as automatic detection
of file delimiter, using <code><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">data.table::fread()</a></code>. This means
that the <code>delim</code> argument has been superseded and it’s often
no longer nessecary to pass the names of the input column names with the
<code>column_names</code> argument. But remember to double check the
logfile that the column names were parsed correctly.</p>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>This is what a typical call to <code><a href="../reference/tidyGWAS.html">tidyGWAS()</a></code> will look
like. A detailed explanation of the different arguments follows.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we setup a directory where we will store all summary statistics cleaned by tidyGWAS</span></span>
<span><span class="va">gwas_folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># provide the filepath to the name of a directory that tidyGWAS will create.</span></span>
<span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">gwas_folder</span>, <span class="st">"gwasName"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="st">"/filepath/to/sumstats"</span>, </span>
<span>  dbsnp_path <span class="op">=</span> <span class="va">dsnp_path</span>,</span>
<span>  <span class="co"># Number of samples are missing, so manually impute</span></span>
<span>  CaseN <span class="op">=</span> <span class="fl">54000</span>,</span>
<span>  ControlN <span class="op">=</span> <span class="fl">73000</span>,</span>
<span>  logfile<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="va">outdir</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="tutorial">Tutorial<a class="anchor" aria-label="anchor" href="#tutorial"></a>
</h2>
<p>tidyGWAS now supports directly downloading files from GWAS catalog or
providing URLs.</p>
<p>The first argument to tidyGWAS is <code>tbl</code>, and should be one
off</p>
<ol style="list-style-type: decimal">
<li>A <a href="https://www.ebi.ac.uk/gwas/" class="external-link">GWAS catalog</a> study id.
For example, “<a href="https://www.ebi.ac.uk/gwas/studies/GCST90101808" class="external-link">GCST90101808</a>”.
The full summary statistics needs to be available.</li>
<li>A webpage URL to a downloadable file. For example, the FinnGEN R<a href="https://storage.googleapis.com/finngen-public-data-r12/summary_stats/release/finngen_R12_F5_DEPRESSIO.gz" class="external-link">12
depression</a> release</li>
<li>A local filepath. For example
“~/summary_stats/raw/trait_x.tsv.gz”</li>
<li>An in-memory data.frame.</li>
</ol>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ararder.github.io/tidyGWAS/">tidyGWAS</a></span><span class="op">)</span></span>
<span><span class="co"># we use the dummy version of dbSNP that comes with the package</span></span>
<span><span class="va">dbsnp_path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/dbSNP155"</span>, package <span class="op">=</span> <span class="st">"tidyGWAS"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># a dummy sumstats with 100 000 rows</span></span>
<span><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sumstats.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"tidyGWAS"</span><span class="op">)</span></span>
<span><span class="co"># store the results in a temporary directory</span></span>
<span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># using a GWAS catalog study ID</span></span>
<span><span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="st">"GCST90101808"</span>,</span>
<span>  dbsnp_path <span class="op">=</span> <span class="va">dbsnp_path</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="st">"finnGEN_MDD_R12"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># using a URL</span></span>
<span><span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="st">"https://storage.googleapis.com/finngen-public-data-r12/summary_stats/release/finngen_R12_F5_DEPRESSIO.gz"</span>,</span>
<span>  dbsnp_path <span class="op">=</span> <span class="va">dbsnp_path</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="st">"finnGEN_MDD_R12"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># using a local flepath</span></span>
<span><span class="va">gwas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sumstats.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"tidyGWAS"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">gwas</span>,</span>
<span>  dbsnp_path <span class="op">=</span> <span class="va">dbsnp_path</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="st">"test_sumstat"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or an in-memory data.frame</span></span>
<span><span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">gwas</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">df</span>,</span>
<span>  dbsnp_path <span class="op">=</span> <span class="va">dbsnp_path</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="st">"test_sumstat"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="an-example-call">An example call<a class="anchor" aria-label="anchor" href="#an-example-call"></a>
</h3>
<p>Here’s what the output of tidyGWAS looks like, using the test file
provided with tidyGWAS</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ararder.github.io/tidyGWAS/">tidyGWAS</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sumstats.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"tidyGWAS"</span><span class="op">)</span>,</span>
<span>  dbsnp_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/dbSNP155"</span>, package <span class="op">=</span> <span class="st">"tidyGWAS"</span><span class="op">)</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Parsing input summary statistics...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Running </span><span style="color: #0000BB; font-weight: bold;">tidyGWAS 1.0.0</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Starting at 2025-08-21 11:43:38.539257</span></span>
<span><span class="co">#&gt; with 100000 rows in input data.frame</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Saving output in folder: <span style="color: #0000BB;">/tmp/RtmpXNz0Nt/file202f26bfe750</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── The detected format is <span style="color: #0000BB;">"tidyGWAS"</span> </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Was able to map 12 out of 12 to the <span style="color: #0000BB;">"tidyGWAS"</span> format</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> CHR -&gt; CHR</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> POS -&gt; POS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> RSID -&gt; RSID</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> EffectAllele -&gt; EffectAllele</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> OtherAllele -&gt; OtherAllele</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> B -&gt; B</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> SE -&gt; SE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> EAF -&gt; EAF</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> P -&gt; P</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> CaseN -&gt; CaseN</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> ControlN -&gt; ControlN</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> INFO -&gt; INFO</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking that columns follow tidyGWAS format </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> The following columns are used for further steps: <span style="font-style: italic;">CHR</span>, <span style="font-style: italic;">POS</span>, <span style="font-style: italic;">RSID</span>, <span style="font-style: italic;">EffectAllele</span>, <span style="font-style: italic;">OtherAllele</span>, <span style="font-style: italic;">B</span>, <span style="font-style: italic;">SE</span>, <span style="font-style: italic;">EAF</span>, <span style="font-style: italic;">INFO</span>, <span style="font-style: italic;">P</span>, <span style="font-style: italic;">CaseN</span>, <span style="font-style: italic;">ControlN</span>, and <span style="font-style: italic;">rowid</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking for columns with all NA </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no columns with all NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Found CaseN and ControlN, and no effective N:</span></span>
<span><span class="co">#&gt; Calculating EffectiveN by `EffectiveN = 4 / (1 / ControlN + 1 / CaseN)`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Found CaseN and ControlN, Calculating N by `N = ControlN + CaseN`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">1) Scanning for rows with NA in critical columns</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> No rows contained missing values in CHR, POS, EffectAllele, and OtherAllele</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">2) Scanning for rows with duplications</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Looking for duplications with columns: CHR, POS, EffectAllele, and OtherAllele</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no duplications</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">3) Scanning for indels</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. EffectAllele or OtherAllele, character length &gt; 1: A vs AA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. EffectAllele or OtherAllele coded as 'D', 'I', or 'R'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Detected 0 rows as indels</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 4) Validating columns </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The median value of B is -0.0012, which seems reasonable</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-style: italic;">All rows passed validation</span> </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 5) Adding RSID based on CHR:POS. Adding dbSNP based QC flags </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Inferring build by matching 10000 rows to GRCh37 and GRCh38</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 99 snps matched GRCh38, 9998 for GRCh37, inferring build to be 37</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> 0 rows had EffectAllele and OtherAllele not matching dbSNP REF/ALT alleles,but which matched after strand-flipping</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> Removed 21 rows with no dbSNP entry or with incompat alleles</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">/tmp/RtmpXNz0Nt/file202f26bfe750/pipeline_info/removed_nodbsnp.parquet</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">6) Repairing missings statistics columns if possible</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Z missing: Calculating Z using the formula:  Z = B / SE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Finished repair_stats:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Added 1 new columns: Z</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Listing final breakdown of removed rows:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; nodbsnp: 21</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Finished tidyGWAS</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A total of 21 rows were removed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Total running time: 3.9s</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Saving metadata from analysis to <span style="color: #0000BB;">/tmp/RtmpXNz0Nt/file202f26bfe750/metadata.yaml</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 99,979 × 21</span></span></span>
<span><span class="co">#&gt;    CHR   POS_37 EffectAllele OtherAllele rowid     B        P     SE  INFO CaseN</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 6     2.75<span style="color: #949494;">e</span>7 C            T           <span style="text-decoration: underline;">77</span>548 0.212 4.83<span style="color: #949494;">e</span><span style="color: #BB0000;">-39</span> 0.016<span style="text-decoration: underline;">2</span> 0.99  <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 6     2.81<span style="color: #949494;">e</span>7 T            C           <span style="text-decoration: underline;">36</span>396 0.207 4.67<span style="color: #949494;">e</span><span style="color: #BB0000;">-38</span> 0.016  1     <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 6     2.75<span style="color: #949494;">e</span>7 C            T           <span style="text-decoration: underline;">77</span>992 0.206 7.92<span style="color: #949494;">e</span><span style="color: #BB0000;">-38</span> 0.016  0.993 <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 6     2.78<span style="color: #949494;">e</span>7 T            C           <span style="text-decoration: underline;">67</span>347 0.171 1.45<span style="color: #949494;">e</span><span style="color: #BB0000;">-32</span> 0.014<span style="text-decoration: underline;">4</span> 1     <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 6     2.63<span style="color: #949494;">e</span>7 T            C            <span style="text-decoration: underline;">7</span>577 0.191 8.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-30</span> 0.016<span style="text-decoration: underline;">8</span> 0.996 <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 6     3.08<span style="color: #949494;">e</span>7 G            A           <span style="text-decoration: underline;">26</span>615 0.157 9.77<span style="color: #949494;">e</span><span style="color: #BB0000;">-28</span> 0.014<span style="text-decoration: underline;">4</span> 0.954 <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 6     2.83<span style="color: #949494;">e</span>7 C            T           <span style="text-decoration: underline;">91</span>092 0.116 4.33<span style="color: #949494;">e</span><span style="color: #BB0000;">-27</span> 0.010<span style="text-decoration: underline;">8</span> 1.01  <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 6     2.64<span style="color: #949494;">e</span>7 C            T           <span style="text-decoration: underline;">84</span>820 0.151 9.52<span style="color: #949494;">e</span><span style="color: #BB0000;">-27</span> 0.014<span style="text-decoration: underline;">1</span> 1     <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 6     3.11<span style="color: #949494;">e</span>7 C            T           <span style="text-decoration: underline;">17</span>683 0.143 1.54<span style="color: #949494;">e</span><span style="color: #BB0000;">-25</span> 0.013<span style="text-decoration: underline;">7</span> 1     <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 6     3.03<span style="color: #949494;">e</span>7 A            G           <span style="text-decoration: underline;">71</span>635 0.154 2.58<span style="color: #949494;">e</span><span style="color: #BB0000;">-25</span> 0.014<span style="text-decoration: underline;">8</span> 0.992 <span style="text-decoration: underline;">49</span>683</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 99,969 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 11 more variables: ControlN &lt;int&gt;, EAF &lt;dbl&gt;, EffectiveN &lt;int&gt;, N &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   POS_38 &lt;int&gt;, RSID &lt;chr&gt;, REF_37 &lt;chr&gt;, REF_38 &lt;chr&gt;, multi_allelic &lt;lgl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   indel &lt;lgl&gt;, Z &lt;dbl&gt;</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="manually-setting-column-names">Manually setting column names<a class="anchor" aria-label="anchor" href="#manually-setting-column-names"></a>
</h3>
<p>tidyGWAS will attempt to guess the column names of the input file.
Always take a look at the logfile if it’s the first time you are using
that particular format. Sometimes the automatic parsing fails, like in
the example below</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testdf</span> <span class="op">&lt;-</span> <span class="fu">arrow</span><span class="fu">::</span><span class="fu"><a href="https://arrow.apache.org/docs/r/reference/read_delim_arrow.html" class="external-link">read_tsv_arrow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sumstats.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"tidyGWAS"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># nonsense names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">testdf</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CHROM_X"</span>, <span class="st">"BP_NEW"</span>, <span class="st">"RsID"</span>, <span class="st">"A1"</span>, <span class="st">"A2"</span>, <span class="st">"B"</span>,<span class="st">"SE"</span>, <span class="st">"EAF"</span>,<span class="st">"INFO"</span>, <span class="st">"P"</span>, <span class="st">"CaseN"</span>, <span class="st">"ControlN"</span>, <span class="st">"rowid"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Wit</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>    tbl <span class="op">=</span> <span class="va">testdf</span>,</span>
<span>    dbsnp_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/dbSNP155"</span>, package <span class="op">=</span> <span class="st">"tidyGWAS"</span><span class="op">)</span>,</span>
<span>    output_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span>,</span>
<span>  error <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">e</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span><span class="op">(</span><span class="st">"tidyGWAS failed: "</span>, <span class="va">e</span><span class="op">$</span><span class="va">message</span><span class="op">)</span></span>
<span>    <span class="cn">NULL</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Parsing input summary statistics...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Running </span><span style="color: #0000BB; font-weight: bold;">tidyGWAS 1.0.0</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Starting at 2025-08-21 11:43:42.620917</span></span>
<span><span class="co">#&gt; with 100000 rows in input data.frame</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Saving output in folder: <span style="color: #0000BB;">/tmp/RtmpXNz0Nt/file202f4ba943fc</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── The detected format is <span style="color: #0000BB;">"tidyGWAS"</span> </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Was able to map 7 out of 12 to the <span style="color: #0000BB;">"tidyGWAS"</span> format</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> B -&gt; B</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> SE -&gt; SE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> EAF -&gt; EAF</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> P -&gt; P</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> CaseN -&gt; CaseN</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> ControlN -&gt; ControlN</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> INFO -&gt; INFO</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> Failed to map: <span style="color: #0000BB;">"CHROM_X"</span>, <span style="color: #0000BB;">"BP_NEW"</span>, <span style="color: #0000BB;">"RsID"</span>, <span style="color: #0000BB;">"A1"</span>, and <span style="color: #0000BB;">"A2"</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Attempting to map: <span style="color: #0000BB;">"CHROM_X"</span>, <span style="color: #0000BB;">"BP_NEW"</span>, <span style="color: #0000BB;">"RsID"</span>, <span style="color: #0000BB;">"A1"</span>, and <span style="color: #0000BB;">"A2"</span> to a dictionary of column names</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> EffectAllele -&gt; A1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> OtherAllele -&gt; A2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> extra column(s) <span style="color: #0000BB;">"CHROM_X"</span>, <span style="color: #0000BB;">"BP_NEW"</span>, and <span style="color: #0000BB;">"RsID"</span> were not mapped to a tidyGWAS column</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking that columns follow tidyGWAS format </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> The following columns are used for further steps: <span style="font-style: italic;">EffectAllele</span>, <span style="font-style: italic;">OtherAllele</span>, <span style="font-style: italic;">B</span>, <span style="font-style: italic;">SE</span>, <span style="font-style: italic;">EAF</span>, <span style="font-style: italic;">INFO</span>, <span style="font-style: italic;">P</span>, <span style="font-style: italic;">CaseN</span>, <span style="font-style: italic;">ControlN</span>, and <span style="font-style: italic;">rowid</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> <span style="font-weight: bold;">Removed columns:  CHROM_X, BP_NEW, and RsID</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking for columns with all NA </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no columns with all NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; tidyGWAS failed: Either CHR and POS or RSID are required columns</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre></div>
<p>If you use column_names, those columns will be renamed, following by
an automatic attempt at parsing the remaining column names.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">testdf</span> <span class="op">&lt;-</span> <span class="fu">arrow</span><span class="fu">::</span><span class="fu"><a href="https://arrow.apache.org/docs/r/reference/read_delim_arrow.html" class="external-link">read_tsv_arrow</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/sumstats.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"tidyGWAS"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># nonsense names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">testdf</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CHROM_X"</span>, <span class="st">"BP_NEW"</span>, <span class="st">"RsID"</span>, <span class="st">"A1"</span>, <span class="st">"A2"</span>, <span class="st">"B"</span>,<span class="st">"SE"</span>, <span class="st">"EAF"</span>,<span class="st">"INFO"</span>, <span class="st">"P"</span>, <span class="st">"CaseN"</span>, <span class="st">"ControlN"</span>, <span class="st">"rowid"</span><span class="op">)</span></span>
<span><span class="va">out</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">testdf</span>,</span>
<span>  dbsnp_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata/dbSNP155"</span>, package <span class="op">=</span> <span class="st">"tidyGWAS"</span><span class="op">)</span>,</span>
<span>  output_dir <span class="op">=</span> <span class="va">out</span>,</span>
<span>  column_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    CHR <span class="op">=</span> <span class="st">"CHROM_X"</span>,</span>
<span>    POS <span class="op">=</span> <span class="st">"BP_NEW"</span>,</span>
<span>    RSID <span class="op">=</span> <span class="st">"RsID"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Parsing input summary statistics...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Running </span><span style="color: #0000BB; font-weight: bold;">tidyGWAS 1.0.0</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Starting at 2025-08-21 11:43:42.975178</span></span>
<span><span class="co">#&gt; with 100000 rows in input data.frame</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Saving output in folder: <span style="color: #0000BB;">/tmp/RtmpXNz0Nt/file202f39f99873</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── The detected format is <span style="color: #0000BB;">"tidyGWAS"</span> </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Was able to map 10 out of 12 to the <span style="color: #0000BB;">"tidyGWAS"</span> format</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> CHR -&gt; CHR</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> POS -&gt; POS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> RSID -&gt; RSID</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> B -&gt; B</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> SE -&gt; SE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> EAF -&gt; EAF</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> P -&gt; P</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> CaseN -&gt; CaseN</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> ControlN -&gt; ControlN</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> INFO -&gt; INFO</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> Failed to map: <span style="color: #0000BB;">"A1"</span> and <span style="color: #0000BB;">"A2"</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Attempting to map: <span style="color: #0000BB;">"A1"</span> and <span style="color: #0000BB;">"A2"</span> to a dictionary of column names</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> EffectAllele -&gt; A1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> OtherAllele -&gt; A2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking that columns follow tidyGWAS format </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> The following columns are used for further steps: <span style="font-style: italic;">CHR</span>, <span style="font-style: italic;">POS</span>, <span style="font-style: italic;">RSID</span>, <span style="font-style: italic;">EffectAllele</span>, <span style="font-style: italic;">OtherAllele</span>, <span style="font-style: italic;">B</span>, <span style="font-style: italic;">SE</span>, <span style="font-style: italic;">EAF</span>, <span style="font-style: italic;">INFO</span>, <span style="font-style: italic;">P</span>, <span style="font-style: italic;">CaseN</span>, <span style="font-style: italic;">ControlN</span>, and <span style="font-style: italic;">rowid</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking for columns with all NA </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no columns with all NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Found CaseN and ControlN, and no effective N:</span></span>
<span><span class="co">#&gt; Calculating EffectiveN by `EffectiveN = 4 / (1 / ControlN + 1 / CaseN)`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Found CaseN and ControlN, Calculating N by `N = ControlN + CaseN`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">1) Scanning for rows with NA in critical columns</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> No rows contained missing values in CHR, POS, EffectAllele, and OtherAllele</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">2) Scanning for rows with duplications</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Looking for duplications with columns: CHR, POS, EffectAllele, and OtherAllele</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no duplications</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">3) Scanning for indels</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. EffectAllele or OtherAllele, character length &gt; 1: A vs AA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. EffectAllele or OtherAllele coded as 'D', 'I', or 'R'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Detected 0 rows as indels</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 4) Validating columns </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The median value of B is -0.0012, which seems reasonable</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-style: italic;">All rows passed validation</span> </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 5) Adding RSID based on CHR:POS. Adding dbSNP based QC flags </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Inferring build by matching 10000 rows to GRCh37 and GRCh38</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 82 snps matched GRCh38, 9997 for GRCh37, inferring build to be 37</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> 0 rows had EffectAllele and OtherAllele not matching dbSNP REF/ALT alleles,but which matched after strand-flipping</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #BBBB00;">!</span> Removed 21 rows with no dbSNP entry or with incompat alleles</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">/tmp/RtmpXNz0Nt/file202f39f99873/pipeline_info/removed_nodbsnp.parquet</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">6) Repairing missings statistics columns if possible</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Z missing: Calculating Z using the formula:  Z = B / SE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Finished repair_stats:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Added 1 new columns: Z</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Listing final breakdown of removed rows:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; nodbsnp: 21</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Finished tidyGWAS</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A total of 21 rows were removed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Total running time: 3.4s</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Saving metadata from analysis to <span style="color: #0000BB;">/tmp/RtmpXNz0Nt/file202f39f99873/metadata.yaml</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 99,979 × 21</span></span></span>
<span><span class="co">#&gt;    CHR   POS_37 EffectAllele OtherAllele rowid     B        P     SE  INFO CaseN</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> 6     2.75<span style="color: #949494;">e</span>7 C            T           <span style="text-decoration: underline;">77</span>548 0.212 4.83<span style="color: #949494;">e</span><span style="color: #BB0000;">-39</span> 0.016<span style="text-decoration: underline;">2</span> 0.99  <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> 6     2.81<span style="color: #949494;">e</span>7 T            C           <span style="text-decoration: underline;">36</span>396 0.207 4.67<span style="color: #949494;">e</span><span style="color: #BB0000;">-38</span> 0.016  1     <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> 6     2.75<span style="color: #949494;">e</span>7 C            T           <span style="text-decoration: underline;">77</span>992 0.206 7.92<span style="color: #949494;">e</span><span style="color: #BB0000;">-38</span> 0.016  0.993 <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> 6     2.78<span style="color: #949494;">e</span>7 T            C           <span style="text-decoration: underline;">67</span>347 0.171 1.45<span style="color: #949494;">e</span><span style="color: #BB0000;">-32</span> 0.014<span style="text-decoration: underline;">4</span> 1     <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> 6     2.63<span style="color: #949494;">e</span>7 T            C            <span style="text-decoration: underline;">7</span>577 0.191 8.94<span style="color: #949494;">e</span><span style="color: #BB0000;">-30</span> 0.016<span style="text-decoration: underline;">8</span> 0.996 <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> 6     3.08<span style="color: #949494;">e</span>7 G            A           <span style="text-decoration: underline;">26</span>615 0.157 9.77<span style="color: #949494;">e</span><span style="color: #BB0000;">-28</span> 0.014<span style="text-decoration: underline;">4</span> 0.954 <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> 6     2.83<span style="color: #949494;">e</span>7 C            T           <span style="text-decoration: underline;">91</span>092 0.116 4.33<span style="color: #949494;">e</span><span style="color: #BB0000;">-27</span> 0.010<span style="text-decoration: underline;">8</span> 1.01  <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> 6     2.64<span style="color: #949494;">e</span>7 C            T           <span style="text-decoration: underline;">84</span>820 0.151 9.52<span style="color: #949494;">e</span><span style="color: #BB0000;">-27</span> 0.014<span style="text-decoration: underline;">1</span> 1     <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> 6     3.11<span style="color: #949494;">e</span>7 C            T           <span style="text-decoration: underline;">17</span>683 0.143 1.54<span style="color: #949494;">e</span><span style="color: #BB0000;">-25</span> 0.013<span style="text-decoration: underline;">7</span> 1     <span style="text-decoration: underline;">53</span>386</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> 6     3.03<span style="color: #949494;">e</span>7 A            G           <span style="text-decoration: underline;">71</span>635 0.154 2.58<span style="color: #949494;">e</span><span style="color: #BB0000;">-25</span> 0.014<span style="text-decoration: underline;">8</span> 0.992 <span style="text-decoration: underline;">49</span>683</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 99,969 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 11 more variables: ControlN &lt;int&gt;, EAF &lt;dbl&gt;, EffectiveN &lt;int&gt;, N &lt;int&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   POS_38 &lt;int&gt;, RSID &lt;chr&gt;, REF_37 &lt;chr&gt;, REF_38 &lt;chr&gt;, multi_allelic &lt;lgl&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   indel &lt;lgl&gt;, Z &lt;dbl&gt;</span></span></span></code></pre></div>
<div class="section level4">
<h4 id="output-files">Output files<a class="anchor" aria-label="anchor" href="#output-files"></a>
</h4>
<p>tidyGWAS returns all its output in a directory.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_tree.html" class="external-link">dir_tree</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span></span>
<span><span class="co">#&gt; /tmp/RtmpXNz0Nt/file202f39f99873</span></span>
<span><span class="co">#&gt; ├── metadata.yaml</span></span>
<span><span class="co">#&gt; ├── pipeline_info</span></span>
<span><span class="co">#&gt; │   ├── removed_nodbsnp.parquet</span></span>
<span><span class="co">#&gt; │   └── removed_validate_chr_pos_path.parquet.parquet</span></span>
<span><span class="co">#&gt; ├── raw</span></span>
<span><span class="co">#&gt; │   └── raw.parquet</span></span>
<span><span class="co">#&gt; └── tidyGWAS_hivestyle</span></span>
<span><span class="co">#&gt;     ├── CHR=1</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=10</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=11</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=12</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=13</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=14</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=15</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=16</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=17</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=18</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=19</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=2</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=20</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=21</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=22</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=3</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=4</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=5</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=6</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=7</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     ├── CHR=8</span></span>
<span><span class="co">#&gt;     │   └── part-0.parquet</span></span>
<span><span class="co">#&gt;     └── CHR=9</span></span>
<span><span class="co">#&gt;         └── part-0.parquet</span></span></code></pre></div>
<p>That’s a lot of files!</p>
<ol style="list-style-type: decimal">
<li><p>Removed rows are stored in
<code>pipeline/removed_*</code></p></li>
<li><p><code>metadata.yaml</code> contains metadata about the
execution</p></li>
<li><p><code>raw</code> contains the summary statistics
<strong><em>before</em></strong> any munging was done. Useful to
reproduce, or to identify why rows were removed.</p></li>
<li><p><code>tidyGWAS_hivestyle</code> contains the cleaned summary
statistics, in something called a <a href="https://arrow.apache.org/docs/r/articles/dataset.html" class="external-link">hivestyle
partition</a>, by default. The motivation for this is detailed further
<a href="#hivestyle-partitioning">down</a> in the vignette. If you just
want a standard csv file, use <code>output_format="csv"</code> or
<code>output_format="parquet"</code></p></li>
<li>
<p>To read in the cleaned summary statistics:</p>
<p><code>df &lt;- arrow::open_dataset(paste0(out, "/tidyGWAS_hivestyle")) |&gt; dplyr::collect()</code></p>
</li>
</ol>
</div>
<div class="section level4">
<h4 id="the-tidygwas-columns">The tidyGWAS columns<a class="anchor" aria-label="anchor" href="#the-tidygwas-columns"></a>
</h4>
<p>tidyGWAS uses the following column names:</p>
<ul>
<li><p>CHR</p></li>
<li><p>POS</p></li>
<li><p>RSID</p></li>
<li><p>EffectAllele</p></li>
<li><p>OtherAllele</p></li>
<li><p>EAF</p></li>
<li><p>B</p></li>
<li><p>SE</p></li>
<li><p>P</p></li>
<li><p>CaseN</p></li>
<li><p>ControlN</p></li>
<li><p>N</p></li>
<li><p>EffectiveN</p></li>
<li><p>INFO</p></li>
<li><p>Z</p></li>
</ul>
<p><strong>Note</strong>:</p>
<p>If your RSID column is in the format CHR:BP:A1:A2, you can still pass
it as an RSID column. tidyGWAS automatically detects and splits apart
the values.</p>
</div>
<div class="section level4">
<h4 id="inputting-sample-size-columns">Inputting sample size columns<a class="anchor" aria-label="anchor" href="#inputting-sample-size-columns"></a>
</h4>
<p>Often, the sample size column is missing from the summary statistics,
and you provide it manually. tidyGWAS has three arguments that can be
used to manually set the sample size if it’s missing from the original
file:</p>
<ul>
<li><p><code>CaseN</code></p></li>
<li><p><code>ControlN</code></p></li>
<li><p><code>N</code></p></li>
</ul>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">sumstats</span>, </span>
<span>  dbsnp_path <span class="op">=</span> <span class="va">dbsnp_path</span>,</span>
<span>  <span class="co"># CaseN, ControlN and N can all be used to set sample size</span></span>
<span>  CaseN <span class="op">=</span> <span class="fl">400</span>,</span>
<span>  ControlN <span class="op">=</span> <span class="fl">800</span>,</span>
<span>  N <span class="op">=</span> <span class="fl">1200</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="reading-in-files">Reading in files<a class="anchor" aria-label="anchor" href="#reading-in-files"></a>
</h4>
<p>If you pass a filepath to tidyGWAS, it will attempt to read in the
file with <code><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">data.table::fread()</a></code>. If you need to pass custom
arguments to fread to make the parsing correct, you can the arguments
through <code>...</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>cleaned <span class="ot">&lt;-</span> <span class="fu">tidyGWAS</span>(</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>  <span class="at">tbl =</span> <span class="st">"filepath/to/gwas/ondisk.csv"</span>,</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>  <span class="co"># fread has argument `skip`. Let's pretend i want to skip the first 100 rows</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>  <span class="at">skip =</span> <span class="fl">100.</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>  <span class="at">dbsnp_path =</span> dbsnp_path,</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="hivestyle-partitioning">Hivestyle-partitioning<a class="anchor" aria-label="anchor" href="#hivestyle-partitioning"></a>
</h2>
<p>The default output format is a hivestyle format using .parquet files.
See for example <a href="https://arrow.apache.org/docs/r/articles/dataset.html" class="external-link">here</a>
for some motivation for this. In essence, this format will significantly
speed up other downstream applications such as meta-analysis, LD
querying and other analyses.</p>
<div class="section level4">
<h4 id="variant-identity">Variant identity<a class="anchor" aria-label="anchor" href="#variant-identity"></a>
</h4>
<p>TidyGWAS will add the following columns dealing with variant ID, in
addition to saving and validating all valid columns in the input
file.</p>
<ol style="list-style-type: decimal">
<li><p><code>CHR</code> Chromosome. The same across both builds</p></li>
<li><p><code>POS_38</code>, <code>POS_37</code> genomic position on
GRCh38 and GRCh37</p></li>
<li><p><code>RSID</code> variant ID from dbSNP</p></li>
<li><p><code>REF_37</code>, <code>REF_38</code> is the reference genome
allele on GRCh38 and GRCh37</p></li>
<li><p><code>multi_allelic</code> is a TRUE/FALSE column that flag rows
that were multi-allelic IN the summary statistics NOT whether there are
multiple alleles in dbSNP. (TRUE corresponds to multi allelic).</p></li>
<li><p><code>rowid</code> maps each row back to the inputted summary
statistics the file in <code>raw</code>, so that any row can be mapped
back to it’s original values.</p></li>
<li><p><code>indel</code> TRUE/FALSE whether the variant is of type
INsertion/DELetion</p></li>
<li><p>All other valid columns in the input summary statistics
file</p></li>
<li><p>If possible <code><a href="../reference/repair_stats.html">repair_stats()</a></code> will add statistics
columns such as <code>B, P, SE, Z</code>, if they are missing and
possible to repair.</p></li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="parallel-computation">Parallel computation<a class="anchor" aria-label="anchor" href="#parallel-computation"></a>
</h2>
<p>tidyGWAS automatically detects the number of cores. In some cases,
for example when running tidyGWAS in a HPC cluster, you might need to
manually set the number of cores, which can be done using the
<code>OMP_NUM_THREADS</code> variable. This should not be a larger
number of cores than what you have requested in your HPC job (in the
example below, the “–cpus-per-task” flag)</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="co">#SBATCH --mem=60gb</span></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co">#SBATCH --time=24:0:00</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task 8</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="bu">export</span> <span class="va">OMP_NUM_THREADS</span><span class="op">=</span>8</span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="va">outdir</span><span class="op">=</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="va">gwas</span><span class="op">=</span><span class="va">$outdir</span>/my_gwas.tsv.gz</span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="va">dbsnp_files</span><span class="op">=</span><span class="st">"dbSNP155"</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">"tidyGWAS(commandArgs(trailingOnly = TRUE)[1],  dbsnp_path = commandArgs(trailingOnly = TRUE)[2],output_dir = commandArgs(trailingOnly = TRUE)[3], logfile=TRUE)"</span> <span class="va">$gwas</span> <span class="va">$dbsnp_files</span> <span class="va">$outdir</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="computational-cost-and-memory-usage">Computational cost and memory usage<a class="anchor" aria-label="anchor" href="#computational-cost-and-memory-usage"></a>
</h2>
<p>Memory use and time scales with the size of the summary statistics.
From running tidyGWAS, here’s an estimation. Especially, if the summary
statistics are missing CHR and POS, tidyGWAS will require additional
memory. 1. 10 million rows ~ 5-15gb 2. 40 million rows 10-40gb 3. 60
million rows 65-85gb</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Arvid Harder.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
