<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="tidyGWAS">
<title>tidyGWAS • tidyGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="tidyGWAS">
<meta property="og:description" content="tidyGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tidyGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.6.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/tidyGWAS.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/transforming_dbsnp_to_parquet.html">get-to-know-dbsnp</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>tidyGWAS</h1>
            
      
      
      <div class="d-none name"><code>tidyGWAS.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<div class="section level4">
<h4 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h4>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ararder/tidyGWAS"</span><span class="op">)</span> <span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ararder/tidyGWAS"</span><span class="op">)</span> <span class="co"># install.packages("remotes")</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="download-reference-files---this-only-has-to-be-done-once">Download reference files - This only has to be done once<a class="anchor" aria-label="anchor" href="#download-reference-files---this-only-has-to-be-done-once"></a>
</h4>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://arvidharder.com/tidyGWAS/">tidyGWAS</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org" class="external-link">fs</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Provide filepath to a directory where you want to store reference files</span></span>
<span><span class="va">ref_dir</span> <span class="op">&lt;-</span> <span class="st">"path/to/storage/directory"</span></span>
<span><span class="fu"><a href="../reference/download_ref_files.html">download_ref_files</a></span><span class="op">(</span><span class="va">ref_dir</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the reference files will be stored in ref_dir/dbSNP155</span></span>
<span><span class="va">dbsnp_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">ref_dir</span>, <span class="st">"/dbSNP155"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="provide-column-names-tidygwas-understands">Provide column names tidyGWAS understands<a class="anchor" aria-label="anchor" href="#provide-column-names-tidygwas-understands"></a>
</h4>
<p>tidyGWAS does NOT automatically parse common column names, so you
need to provide it with the correct column names.
<code><a href="../reference/tidyGWAS_columns.html">tidyGWAS_columns()</a></code> is a helper function to do this.
tidyGWAS uses the following naming:</p>
<p>Note: If your RSID column is in the format CHR:BP:A1:A2, you can
still pass it as RSID. Note: If N is missing from your sumstats, you can
pass that later in [tidyGWAS()] The tidyGWAS column names: CHR POS RSID
EffectAllele OtherAllele EAF B SE P CaseN ControlN N INFO OR Z</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read in the sumstats you want to clean</span></span>
<span><span class="co"># sumstats &lt;- readr::read_tsv("path/to_sumstats/gwas.tsv.gz")</span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS_columns.html">tidyGWAS_columns</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">sumstats</span>,</span>
<span>  CHR <span class="op">=</span> <span class="st">"CHROM"</span>,</span>
<span>  POS <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>  EffectAllele <span class="op">=</span> <span class="st">"Allele1"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">df</span>, </span>
<span>  <span class="co"># provide the filepath to the refence files you downloaded previously</span></span>
<span>  <span class="co"># and add dbSNP155</span></span>
<span>  dbsnp_path <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">dbsnp_files</span>, <span class="st">"/dbSNP155"</span><span class="op">)</span>,</span>
<span>  <span class="co"># it's useful to write all the messages to a logfile, so messages are saved.</span></span>
<span>  logfile<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>  output_format <span class="op">=</span> <span class="st">"csv"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="in-depth-walk-through">In depth walk-through<a class="anchor" aria-label="anchor" href="#in-depth-walk-through"></a>
</h2>
<p>Let’s walk through each step more carefully, using some example files
provided in the tidyGWAS package. We will be using a version of the
sumstats from the latest GWAS on schizophrenia from the Psychiatric
Genomics Consortium <a href="https://www.nature.com/articles/s41586-022-04434-5" class="external-link">(PGC3)</a></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://arvidharder.com/tidyGWAS/">tidyGWAS</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org" class="external-link">fs</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">example_sumstats</span> <span class="op">&lt;-</span> <span class="fu">tidyGWAS</span><span class="fu">::</span><span class="va"><a href="../reference/test_file.html">test_file</a></span></span>
<span><span class="co"># here we use the example files provided in the tidyGWAS package</span></span>
<span><span class="va">example_ref_file</span> <span class="op">&lt;-</span>  <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"tidyGWAS"</span><span class="op">)</span>, <span class="st">"extdata/dbSNP155"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#The columns have been renamed to follow the tidyGWAS format:</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">example_sumstats</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "CHR"          "POS"          "RSID"         "EffectAllele" "OtherAllele" </span></span>
<span><span class="co">#&gt;  [6] "B"            "SE"           "EAF"          "INFO"         "P"           </span></span>
<span><span class="co">#&gt; [11] "CaseN"        "ControlN"</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_sumstats</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   CHR      POS        RSID EffectAllele OtherAllele            B     SE</span></span>
<span><span class="co">#&gt; 1   6 31819813 rs183988210            C           G -0.005696193 0.0292</span></span>
<span><span class="co">#&gt; 2  16 13991338   rs8044430            T           C -0.017695649 0.0092</span></span>
<span><span class="co">#&gt; 3   9 85016930 rs111404043            A           T -0.002302649 0.0185</span></span>
<span><span class="co">#&gt; 4  19 15743403   rs4807961            C           G  0.002397125 0.0099</span></span>
<span><span class="co">#&gt; 5  18  5948313    rs948293            T           C  0.001698557 0.0092</span></span>
<span><span class="co">#&gt; 6  13 47224772       rs837            G           A  0.027196792 0.0086</span></span>
<span><span class="co">#&gt;         EAF  INFO        P CaseN ControlN</span></span>
<span><span class="co">#&gt; 1 0.9754088 0.936 0.844400 51235    74087</span></span>
<span><span class="co">#&gt; 2 0.6407741 0.951 0.053150 53386    77258</span></span>
<span><span class="co">#&gt; 3 0.9160000 0.701 0.899700 53386    77258</span></span>
<span><span class="co">#&gt; 4 0.2556345 0.996 0.807400 53386    77258</span></span>
<span><span class="co">#&gt; 5 0.3318173 0.971 0.857800 53386    77258</span></span>
<span><span class="co">#&gt; 6 0.4470432 1.010 0.001594 53386    77258</span></span></code></pre></div>
<p>The only mandatory argument is <code>tbl</code>. tidyGWAS accepts
either a in-memory <code><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">base::data.frame()</a></code> or a filepath to text
file. If you pass a filepath to a text file, tidyGWAS will attempt to
read it using <code><a href="https://arrow.apache.org/docs/r/reference/read_delim_arrow.html" class="external-link">arrow::read_tsv_arrow()</a></code>.</p>
<p>It’s not uncommon for sumstats to have formats that is not supported
by this function, in which case i recommend you read in the file using
<code>data.table::fread()</code> or <code>readr::read_tsv()</code>
first, and then pass the <code>data.frame</code> to
<code><a href="../reference/tidyGWAS.html">tidyGWAS()</a></code>.</p>
<p>if <code>dbsnp_path</code> is not passed, tidyGWAS will skip any
functions relating to dbSNP, such as repairing CHR or POS, or detecting
and providing coordinate on both genome builds.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a minimal example, without dbSNP reference files</span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">example_sumstats</span> </span>
<span>  <span class="co"># dbsnp_path = dbsnp_file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Passing a filepath also works</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>cleaned <span class="ot">&lt;-</span> <span class="fu">tidyGWAS</span>(</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">tbl =</span> <span class="st">"/home/all_sumstats/raw_sumstats.txt.gz"</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">dbsnp_path =</span> dbsnp_file</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># if the file is not parsed correctly using arrow::read_tsv_arrow, you can use</span></span>
<span><span class="co"># data.table::fread() or readr::read_tsv() first, and then pass the data.frame to tidyGWAS  </span></span>
<span><span class="va">df</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu">fread</span><span class="op">(</span><span class="st">"/home/all_sumstats/raw_sumstats.txt.gz"</span><span class="op">)</span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">df</span>,</span>
<span>  dbsnp_path <span class="op">=</span> <span class="va">dbsnp_file</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="a-more-typical-workflow">A more typical workflow<a class="anchor" aria-label="anchor" href="#a-more-typical-workflow"></a>
</h4>
<p>Typically, you will call <code><a href="../reference/tidyGWAS.html">tidyGWAS()</a></code> with a few more
arguments than the example above. By default <code><a href="../reference/tidyGWAS.html">tidyGWAS()</a></code>
will use <code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">base::tempdir()</a></code> to write to during execution. In
addition, the cleaned sumstat is also returned by the function. This
means that files will NOT be saved after you quit the R session.<br>
This behavior can be adjusted with the <code>outdir</code> argument,
where files will be copied to <code>outdir</code> after a successful
run.</p>
<p>Let’s take a look at how a more typical call to tidyGWAS would
look:</p>
<ol style="list-style-type: decimal">
<li>
<code>logfile</code> can be set to TRUE, so the output of tidyGWAS
will be saved in a logfile</li>
<li>
<code>name</code> The output directory can be named as a
convenience.</li>
<li>
<code>output_format</code> tidyGWAS supports three output formats:
‘csv’, ‘parquet’ or ‘hivestyle’</li>
<li>
<code>outdir</code> a filepath to a directory where you want to
store the output. This folder should not exist before running tidyGWAS.
If it does, tidyGWAS will throw an error to prevent overwriting of
files</li>
</ol>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># your own sumstat</span></span>
<span><span class="va">example_file</span> <span class="op">&lt;-</span> <span class="fu">tidyGWAS</span><span class="fu">::</span><span class="va"><a href="../reference/test_file.html">test_file</a></span></span>
<span></span>
<span><span class="co"># a directory where you want your data to be stored after succesful cleaning.</span></span>
<span><span class="co"># here we are using a tempdir, as an example</span></span>
<span><span class="va">gwas_folder</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"tidyGWAS"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">example_file</span>, </span>
<span>  dbsnp_path <span class="op">=</span> <span class="va">example_ref_file</span>,</span>
<span>  <span class="co"># it's useful to write all the messages to a logfile, so messages are saved.</span></span>
<span>  logfile<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>  <span class="co"># by default, the name of the output folder is a concotonated call to Sys.time()</span></span>
<span>  <span class="co"># but you can adjust this by using the name argument</span></span>
<span>  <span class="co"># by default, the saved file is written out as a gzipped csv file,</span></span>
<span>  <span class="co"># however, the arrow parquet format offers many advantages over the classical csv file</span></span>
<span>  output_format <span class="op">=</span> <span class="st">"parquet"</span>,</span>
<span>  <span class="co"># by passing the argument outdir, you can tell tidyGWAS to copy over files</span></span>
<span>  <span class="co"># after a finished execution to that directory</span></span>
<span>  outdir <span class="op">=</span> <span class="va">gwas_folder</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Output is redirected to logfile: <span style="color: #0000BB;">/tmp/RtmpiVCy2M/tidyGWAS1a463bf4260c/tidyGWAS_logfile.txt</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Running </span><span style="color: #0000BB; font-weight: bold;">tidyGWAS 0.6.1</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; Starting at 2023-12-02 16:43:37.471175, with 100000 rows in input data.frame</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Saving output in folder: <span style="color: #0000BB;">/tmp/RtmpiVCy2M/tidyGWAS1a463bf4260c</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking that columns follow tidyGWAS format </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> The following columns are used for further steps: <span style="font-style: italic;">CHR</span>, <span style="font-style: italic;">POS</span>, <span style="font-style: italic;">RSID</span>, <span style="font-style: italic;">EffectAllele</span>, <span style="font-style: italic;">OtherAllele</span>, <span style="font-style: italic;">B</span>, <span style="font-style: italic;">SE</span>, <span style="font-style: italic;">EAF</span>, <span style="font-style: italic;">INFO</span>, <span style="font-style: italic;">P</span>, <span style="font-style: italic;">CaseN</span>, <span style="font-style: italic;">ControlN</span>, and <span style="font-style: italic;">rowid</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking for columns with all NA </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no columns with all NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">1) Scanning for rows with NA</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> No rows contained missing values</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">3) Scanning for rows with duplications</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A unique ID is formed by concontenating CHR, POS, EffectAllele, and OtherAllele</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no duplications</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">4) Scanning for indels</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. EffectAllele or OtherAllele, character length &gt; 1: A vs AA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. EffectAllele or OtherAllele coded as 'D', 'I', or 'R'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Detected 0 rows as indels</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 6) Validating columns: for main_rows </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The median value of B is -0.0012, which seems reasonable</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-style: italic;">All rows passed validation</span> </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Inferring build by matching 10000 rows to GRCh37 and GRCh38</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3349 snps matched GRCh38, 9999 for GRCh37, inferring build to be 37</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">8) Repairing missings columns if possible</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Z missing: Calculating Z using the formula:  Z = B / SE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Finished repair_stats:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Added 1 new columns: Z</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Detected 0 rows as indels</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Listing final breakdown of removed rows:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Finished tidyGWAS</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A total of 0 rows were removed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Total running time: 4.2s</span></span></code></pre></div>
<p>If we take a look in the output folder we provided, we can see that a
new folder has been created, we can see the standard files that tidyGWAS
will output</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">gwas_folder</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "cleaned_GRCh38.parquet"                 </span></span>
<span><span class="co">#&gt; [2] "pipeline_info/removed_rows_main.parquet"</span></span>
<span><span class="co">#&gt; [3] "raw_sumstats.parquet"                   </span></span>
<span><span class="co">#&gt; [4] "tidyGWAS_logfile.txt"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="creating-the-correct-column-names">Creating the correct column names<a class="anchor" aria-label="anchor" href="#creating-the-correct-column-names"></a>
</h3>
<p>tidyGWAS does not automatically detect column names, but expect the
<code><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame()</a></code> to contain the correct column names.
<code>tidyGWAS_columns</code> is a helper function created to match
column name to the tidyGWAS format.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># what if we have a file where POS was named bp, and B was named BETA?</span></span>
<span><span class="co"># tidyGWAS will not recognize these as valid column names and would drop them.</span></span>
<span><span class="va">wrong_colnames</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span><span class="va">example_file</span>, bp <span class="op">=</span> <span class="va">POS</span>, BETA <span class="op">=</span> <span class="va">B</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">wrong_colnames</span><span class="op">)</span></span>
<span><span class="co">#&gt;   CHR       bp        RSID EffectAllele OtherAllele         BETA     SE</span></span>
<span><span class="co">#&gt; 1   6 31819813 rs183988210            C           G -0.005696193 0.0292</span></span>
<span><span class="co">#&gt; 2  16 13991338   rs8044430            T           C -0.017695649 0.0092</span></span>
<span><span class="co">#&gt; 3   9 85016930 rs111404043            A           T -0.002302649 0.0185</span></span>
<span><span class="co">#&gt; 4  19 15743403   rs4807961            C           G  0.002397125 0.0099</span></span>
<span><span class="co">#&gt; 5  18  5948313    rs948293            T           C  0.001698557 0.0092</span></span>
<span><span class="co">#&gt; 6  13 47224772       rs837            G           A  0.027196792 0.0086</span></span>
<span><span class="co">#&gt;         EAF  INFO        P CaseN ControlN</span></span>
<span><span class="co">#&gt; 1 0.9754088 0.936 0.844400 51235    74087</span></span>
<span><span class="co">#&gt; 2 0.6407741 0.951 0.053150 53386    77258</span></span>
<span><span class="co">#&gt; 3 0.9160000 0.701 0.899700 53386    77258</span></span>
<span><span class="co">#&gt; 4 0.2556345 0.996 0.807400 53386    77258</span></span>
<span><span class="co">#&gt; 5 0.3318173 0.971 0.857800 53386    77258</span></span>
<span><span class="co">#&gt; 6 0.4470432 1.010 0.001594 53386    77258</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we can use tidyGWAS_columns to specify current names:</span></span>
<span><span class="va">correct_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS_columns.html">tidyGWAS_columns</a></span><span class="op">(</span></span>
<span>  <span class="co"># first argument is the data.frame</span></span>
<span>  tbl <span class="op">=</span> <span class="va">wrong_colnames</span>, </span>
<span>  <span class="co"># then any column which is not correctly named:</span></span>
<span>  POS <span class="op">=</span> <span class="st">"bp"</span>,</span>
<span>  B <span class="op">=</span> <span class="st">"BETA"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">correct_names</span><span class="op">)</span></span>
<span><span class="co">#&gt;   CHR      POS        RSID EffectAllele OtherAllele            B     SE</span></span>
<span><span class="co">#&gt; 1   6 31819813 rs183988210            C           G -0.005696193 0.0292</span></span>
<span><span class="co">#&gt; 2  16 13991338   rs8044430            T           C -0.017695649 0.0092</span></span>
<span><span class="co">#&gt; 3   9 85016930 rs111404043            A           T -0.002302649 0.0185</span></span>
<span><span class="co">#&gt; 4  19 15743403   rs4807961            C           G  0.002397125 0.0099</span></span>
<span><span class="co">#&gt; 5  18  5948313    rs948293            T           C  0.001698557 0.0092</span></span>
<span><span class="co">#&gt; 6  13 47224772       rs837            G           A  0.027196792 0.0086</span></span>
<span><span class="co">#&gt;          P       EAF CaseN ControlN  INFO</span></span>
<span><span class="co">#&gt; 1 0.844400 0.9754088 51235    74087 0.936</span></span>
<span><span class="co">#&gt; 2 0.053150 0.6407741 53386    77258 0.951</span></span>
<span><span class="co">#&gt; 3 0.899700 0.9160000 53386    77258 0.701</span></span>
<span><span class="co">#&gt; 4 0.807400 0.2556345 53386    77258 0.996</span></span>
<span><span class="co">#&gt; 5 0.857800 0.3318173 53386    77258 0.971</span></span>
<span><span class="co">#&gt; 6 0.001594 0.4470432 53386    77258 1.010</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="parallel-computation">Parallel computation<a class="anchor" aria-label="anchor" href="#parallel-computation"></a>
</h2>
<p>tidGWAS automatically detects the number of cores, and uses it to
parallelize the alignment with dbSNP. In some cases, for example when
running tidyGWAS in a HPC cluster, you might need to manually set the
number of cores, which can be done using the
<code>OMP_NUM_THREADS</code> variable. This should not be a larger
number of cores than what you have requested in your HPC job (in the
example below, the “–cpus-per-task” flag)</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mem=60gb</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --time=24:0:00</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task 8</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">OMP_NUM_THREADS</span><span class="op">=</span>8</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="va">outdir</span><span class="op">=</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="va">gwas</span><span class="op">=</span><span class="va">$outdir</span>/my_gwas.tsv.gz</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a><span class="va">dbsnp_files</span><span class="op">=</span><span class="st">"dbSNP155"</span></span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">"tidyGWAS(commandArgs(trailingOnly = TRUE)[1],  dbsnp_path = commandArgs(trailingOnly = TRUE)[2],outdir = commandArgs(trailingOnly = TRUE)[3], logfile=TRUE)"</span> <span class="va">$gwas</span> <span class="va">$dbsnp_files</span> <span class="va">$outdir</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="computational-cost">Computational cost<a class="anchor" aria-label="anchor" href="#computational-cost"></a>
</h2>
<p>Memory use and time scales with the size of the summary statistics.
From running tidyGWAS experimentally, here’s an estimation of time and
memory.</p>
<div class="section level5">
<h5 id="cpu-usage">CPU usage<a class="anchor" aria-label="anchor" href="#cpu-usage"></a>
</h5>
<p>CPU usage depends heavily on CPU type and number of available
cores.</p>
<ol style="list-style-type: decimal">
<li>Macbook Pro M2: 8 million rows, ~4.42 minutes, &lt; 16GB memory
used</li>
<li>AMD EPYC 75F3 32-Core Processor; ~ 6 minutes &lt; 16GB memory
used</li>
</ol>
<p>For 60 million rows:</p>
<ol style="list-style-type: decimal">
<li>AMD EPYC 75F3 32-Core Processor; ~ 34 minutes &lt; 75gb using 5
cores</li>
</ol>
</div>
<div class="section level4">
<h4 id="memory-usage-and-number-of-rows">Memory usage and number of rows<a class="anchor" aria-label="anchor" href="#memory-usage-and-number-of-rows"></a>
</h4>
<p>The memory usage will likely be the main constraint with increasing
size of summary statistics.</p>
<ol style="list-style-type: decimal">
<li>20 million rows &lt; 20gb</li>
<li>40 million rows &lt; 40gb</li>
<li>60 million rows &lt; 75gb</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="download_ref_data-is-not-working">download_ref_data() is not working<a class="anchor" aria-label="anchor" href="#download_ref_data-is-not-working"></a>
</h2>
<p>tidyGWAS uses a version dbSNP 155 converted to the <a href="https://arrow.apache.org" class="external-link">Apache Arrow</a> .parquet files. You can
download the dbSNP155 reference file from inside R using the
<code>googledrive</code> package, or by manually navigating to this <a href="https://drive.google.com/file/d/1LmgfmTQaWwJaFpBHcRQIY_kwe5iN7Pj6/view?usp=share_link" class="external-link">file</a></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># You can download the file from inside R using the googledrive package:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://googledrive.tidyverse.org" class="external-link">googledrive</a></span><span class="op">)</span></span>
<span><span class="fu">googledrive</span><span class="fu">::</span><span class="fu">drive_deauth</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu">googledrive</span><span class="fu">::</span><span class="fu">as_id</span><span class="op">(</span><span class="st">"1LmgfmTQaWwJaFpBHcRQIY_kwe5iN7Pj6"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##### EDIT THIS:</span></span>
<span><span class="va">filepath_to_store_dir</span> <span class="op">&lt;-</span> <span class="st">""</span></span>
<span><span class="co">##### ---------------------</span></span>
<span></span>
<span><span class="fu">drive_download</span><span class="op">(</span><span class="va">id</span>, path <span class="op">=</span> <span class="va">filepath_to_store_dir</span><span class="op">)</span></span></code></pre></div>
<p>The file needs to be untarred</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># change directory to where the downloaded file is</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xf</span> dbSNP155.tar dbSNP155</span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Arvid Harder.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
