<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="tidyGWAS">
<title>tidyGWAS • tidyGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="tidyGWAS">
<meta property="og:description" content="tidyGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tidyGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/tidyGWAS.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>tidyGWAS</h1>
            
      
      
      <div class="d-none name"><code>tidyGWAS.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://arvidharder.com/tidyGWAS/">tidyGWAS</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>tidyGWAS contains some wrapper functions around BSgenome and the <a href="https://bioconductor.org/packages/release/data/annotation/html/SNPlocs.Hsapiens.dbSNP155.GRCh38.html" class="external-link">SNPlocs</a>
bioconductor packages. These are quite large dependencies, almost ~14GB,
and are therefore listed as “suggests”.</p>
<p>If you want to use be able to infer missing RSID, or CHR:POS, these
packages are required.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ararder/tidyGWAS"</span><span class="op">)</span></span>
<span><span class="co"># or</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ararder/tidyGWAS"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># if you want to be able to repair CHR:POS or RSID, you need to download the following</span></span>
<span><span class="co"># bioconductor packages:</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">"BiocManager"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.16"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"BSgenome"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"Biostrings"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"GenomicRanges"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.1000genomes.hs37d5"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"BSgenome.Hsapiens.NCBI.GRCh38"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"SNPlocs.Hsapiens.dbSNP155.GRCh37"</span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"SNPlocs.Hsapiens.dbSNP155.GRCh38"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>Automatic parsing of column names has not been implemented (and
probable never will), so you will have to rename the input dataframe to
the tidyGWAS format. The following column names are accepted.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">snp_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CHR"</span>, <span class="st">"POS"</span>, <span class="st">"RSID"</span>, <span class="st">"EffectAllele"</span>, <span class="st">"OtherAllele"</span>, <span class="st">"rowid"</span><span class="op">)</span></span>
<span><span class="va">info_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"INFO"</span>, <span class="st">"N"</span>, <span class="st">"CaseN"</span>, <span class="st">"ControlN"</span>, <span class="st">"EAF"</span><span class="op">)</span></span>
<span><span class="va">stats_cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B"</span>, <span class="st">"Z"</span>, <span class="st">"OR"</span>, <span class="st">"P"</span>, <span class="st">"SE"</span><span class="op">)</span></span></code></pre></div>
<p>How the example file looks. The mandatory columns are either RSID, or
CHR and POS</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">test_file</span> <span class="op">&lt;-</span> <span class="fu">tidyGWAS</span><span class="fu">::</span><span class="va"><a href="../reference/test_file.html">test_file</a></span></span>
<span><span class="va">test_file</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 100,000 × 12</span></span></span>
<span><span class="co">#&gt;      CHR       POS RSID     EffectAllele OtherAllele        B     SE   EAF  INFO</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     6  31<span style="text-decoration: underline;">819</span>813 rs18398… C            G           -<span style="color: #BB0000;">0.005</span><span style="color: #BB0000; text-decoration: underline;">70</span> 0.029<span style="text-decoration: underline;">2</span> 0.975 0.936</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>    16  13<span style="text-decoration: underline;">991</span>338 rs80444… T            C           -<span style="color: #BB0000;">0.017</span><span style="color: #BB0000; text-decoration: underline;">7</span>  0.009<span style="text-decoration: underline;">2</span> 0.641 0.951</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     9  85<span style="text-decoration: underline;">016</span>930 rs11140… A            T           -<span style="color: #BB0000;">0.002</span><span style="color: #BB0000; text-decoration: underline;">30</span> 0.018<span style="text-decoration: underline;">5</span> 0.916 0.701</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>    19  15<span style="text-decoration: underline;">743</span>403 rs48079… C            G            0.002<span style="text-decoration: underline;">40</span> 0.009<span style="text-decoration: underline;">9</span> 0.256 0.996</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>    18   5<span style="text-decoration: underline;">948</span>313 rs948293 T            C            0.001<span style="text-decoration: underline;">70</span> 0.009<span style="text-decoration: underline;">2</span> 0.332 0.971</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>    13  47<span style="text-decoration: underline;">224</span>772 rs837    G            A            0.027<span style="text-decoration: underline;">2</span>  0.008<span style="text-decoration: underline;">6</span> 0.447 1.01 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>    10  23<span style="text-decoration: underline;">782</span>010 rs79100… A            T           -<span style="color: #BB0000;">0.022</span><span style="color: #BB0000; text-decoration: underline;">6</span>  0.009<span style="text-decoration: underline;">1</span> 0.666 0.995</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>    12  14<span style="text-decoration: underline;">061</span>867 rs99718… A            T            0.007<span style="text-decoration: underline;">80</span> 0.009  0.641 0.999</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     7  57<span style="text-decoration: underline;">393</span>801 rs48706… A            G           -<span style="color: #BB0000;">0.008</span><span style="color: #BB0000; text-decoration: underline;">00</span> 0.009<span style="text-decoration: underline;">7</span> 0.703 0.942</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>     2 222<span style="text-decoration: underline;">786</span>280 rs824920 T            G            0.002<span style="text-decoration: underline;">10</span> 0.012  0.150 0.996</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 99,990 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 3 more variables: P &lt;dbl&gt;, CaseN &lt;int&gt;, ControlN &lt;int&gt;</span></span></span></code></pre></div>
<p>The <a href="https://bioconductor.org/packages/release/bioc/html/BSgenome.html" class="external-link">BSgenome</a>
and <a href="https://bioconductor.org/packages/release/data/annotation/html/SNPlocs.Hsapiens.dbSNP155.GRCh38.html" class="external-link">SNPlocs</a>
packages are used to store and query dbSNP and fasta files to get CHR,
POS, RSID, REF and ALT for all SNPs in dbSNP v155. The first query to
this dataframe will always be slower than subsequent queries, so
whenever you want to munge several summary statistics, it’s get to first
load the points to the bsgenome and snplocs data, which can be done with
<code>get_bsgenome().</code></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bsgenome</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_bsgenome.html">get_bsgenome</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>You can skip the step to align with dbSNP, which will significantly
speed up running time and reduce memory usage, by not providing
<code>bsgnome_objects</code>. To update the RSID colum, you also need to
provide an RsMergeArch file. tidyGWAS will attempt to read in this file
using a filepath provided in the R enviroment “rs_merge_arch”.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># You can download the rs_merge_arch file here:</span></span>
<span><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu">googledrive</span><span class="fu">::</span><span class="fu">as_dribble</span><span class="op">(</span><span class="st">"https://drive.google.com/file/d/1HaES_q8m6rSDDLXvYr0TjQ-ZlyhdZbFX/view?usp=share_link"</span><span class="op">)</span></span>
<span><span class="fu">googledrive</span><span class="fu">::</span><span class="fu">drive_download</span><span class="op">(</span><span class="va">file</span><span class="op">)</span></span>
<span><span class="co"># use usethis to set the env variable rs_merge_arch to the filepath to the downloaded file </span></span>
<span><span class="co"># usethis::edit_r_environ()</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># use the small example file stored with tidyGWAS package</span></span>
<span><span class="va">rs_merge_arch</span> <span class="op">&lt;-</span> <span class="fu">tidyGWAS</span><span class="fu">::</span><span class="va"><a href="../reference/rs_merge_arch.html">rs_merge_arch</a></span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">start</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span><span class="va">test_file</span>, rs_merge_arch <span class="op">=</span> <span class="va">rs_merge_arch</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Running </span><span style="color: #0000BB; font-weight: bold;">tidyGWAS 0.1.0</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; Starting at 2023-08-01 18:13:34.136512</span></span>
<span><span class="co">#&gt; Writing out raw sumstats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Performing initial checks on input data: </span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Columns: CHR, POS, RSID, EffectAllele, OtherAllele, B, SE, EAF, INFO, P, CaseN, and ControlN</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Examining data for missing values across all columns </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   • Using tidyr::drop_na()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no rows with missing values</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking for insertions/deletions ('indels') using:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   1. EffectAllele or OtherAllele, character length &gt; 1: A vs AA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   2. EffectAllele or OtherAllele coded as 'D', 'I', or 'R'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Detected 0 rows as indels</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Updating RSIDs that have been merged using RsMergeArch </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Updated RSID in 175 rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Finished initial checks</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   • 175 rows with updated RSID:</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">/tmp/RtmpdpV3FK/2023_08_01_18_13_34.099964/2023_08_01_18_13_34.099964/pipeline_info/updated_rsid.log.gz</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Starting validation of CHR,POS, RSID, EffectAllele and OtherAllele</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the RSID column: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; • Checking that RSID follows rs format: [rR][sS][1-10]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; • If rows fail rs format, look for CHR:POS or CHR:POS:REF:ALT format</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Found 307 rows with invalid RSID format: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Attempting to parse format...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Succeded in parsing 307 out of 307 rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Parsed format:</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 5 × 5</span></span></span>
<span><span class="co">#&gt;   RSID            CHR   POS       EffectAllele OtherAllele</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 13:64127045_C_A 13    64127045  C            A          </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 2:79900650_T_C  2     79900650  T            C          </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 6:102035530_T_G 6     102035530 T            G          </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 15:24480820_G_C 15    24480820  G            C          </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 11:14359274_C_A 11    14359274  C            A</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Found CHR and POS in those rows where parsing of RSID failing. Will use CHR and POS columns</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating CHR </span></span>
<span><span class="co">#&gt; 1. CHR coerced to character `base::as.character()`</span></span>
<span><span class="co">#&gt; 2. converted to uppercase `stringr::str_to_upper()`</span></span>
<span><span class="co">#&gt; 3. 'CHR' removed (handles UCSC style CHR format)</span></span>
<span><span class="co">#&gt; 4. Check for values outside 1:22, X, Y</span></span>
<span><span class="co">#&gt; 5. Check for NA</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass chr validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating POS </span></span>
<span><span class="co">#&gt; 1. POS coerced to integer `base::as.integer()`</span></span>
<span><span class="co">#&gt; 2. Checks for NA, NaN, Inf</span></span>
<span><span class="co">#&gt; 3. Checks for POS &lt;= 0</span></span>
<span><span class="co">#&gt; 4. Check for POS &gt;= 10^9</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass pos validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating EffectAllele and OtherAllele: </span></span>
<span><span class="co">#&gt; 1. Will error if not type = character</span></span>
<span><span class="co">#&gt; 2. EffectAllele and OtherAllele converted to uppercase</span></span>
<span><span class="co">#&gt; 3. Check for EA/OA values that are not A,C,G or T</span></span>
<span><span class="co">#&gt; 4. Check for NA</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass EA/OA validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Listing how many rows are removed per flag: </span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; invalid_rsid: 307</span></span>
<span><span class="co">#&gt; invalid_chr: 0</span></span>
<span><span class="co">#&gt; invalid_pos: 0</span></span>
<span><span class="co">#&gt; invalid_ea_oa: 0</span></span>
<span><span class="co">#&gt; Removed a total of 307 rows:</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">/tmp/RtmpdpV3FK/2023_08_01_18_13_34.099964/2023_08_01_18_13_34.099964/pipeline_info/validate_snps.log.gz</span></span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Validating statistics columns: B, P, SE, INFO, N, CaseN, ControlN, and EAF</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the EAF column: </span></span>
<span><span class="co">#&gt; 1. Will coerce to double: base::as.double()</span></span>
<span><span class="co">#&gt; 2. Check for EAF &lt;= 0 &amp; EAF &gt;= 1</span></span>
<span><span class="co">#&gt; 3. Checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass EAF validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the SE column: </span></span>
<span><span class="co">#&gt; 1. Will coerce to double: base::as.double()</span></span>
<span><span class="co">#&gt; 2. Check for SE &lt;= 0</span></span>
<span><span class="co">#&gt; 3. Checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass SE validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the P column: </span></span>
<span><span class="co">#&gt; 1. Coerce to double `base::as.double()`</span></span>
<span><span class="co">#&gt; 2. If P == 0, P is converted to `2.225074e-308`. Adjust by passing convert_p =</span></span>
<span><span class="co">#&gt; 'your_preferred_value'</span></span>
<span><span class="co">#&gt; 3. checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; 4. checks that P is within the range: P &gt;= 0 &amp; P &gt; 1</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass P validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the B column: </span></span>
<span><span class="co">#&gt; 1. Will coerce to double: base::as.double()</span></span>
<span><span class="co">#&gt; 2. Check for mislabelled OR by looking at median value</span></span>
<span><span class="co">#&gt; 3. checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; 4. Print median value</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The median value of B is -0.00120072057651888, which seems reasonable</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass B validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the N column: </span></span>
<span><span class="co">#&gt; 1. Will coerce to integer: base::as.integer()</span></span>
<span><span class="co">#&gt; 2. Check for N &lt;= 0</span></span>
<span><span class="co">#&gt; 3. Checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass N validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Listing how many rows are removed per flag: </span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> No rows were removed by these flags</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Finished tidyGWAS</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A total of 307 rows were removed. Started with: 100000 rows, ended with: 99693 rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Listing final breakdown of removed rows:  </span></span>
<span><span class="co">#&gt; validate_snps: 307</span></span>
<span><span class="va">end</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>About 3 seconds for 100k rows</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">end</span> <span class="op">-</span> <span class="va">start</span></span>
<span><span class="co">#&gt; Time difference of 2.364134 secs</span></span></code></pre></div>
<p>If you want the information to be stored in a logfile on disk, you
can use the logfile flag. Note that by default, tidyGWAS uses a folder
in <code>tempdir</code> to store all intermediate files. These files are
not saved indefinitely, so to store the output data, you can also
provide an output directory where you want the copy over the finished
tidyGWAS oflder</p>
<pre><code><span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Output is redirected to logfile: <span style="color: #0000BB;">/tmp/RtmpdpV3FK/2023_08_01_18_13_36.612778/2023_08_01_18_13_36.612778/tidyGWAS_logfile.txt</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Running </span><span style="color: #0000BB; font-weight: bold;">tidyGWAS 0.1.0</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; Starting at 2023-08-01 18:13:36.635084</span></span>
<span><span class="co">#&gt; Writing out raw sumstats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Performing initial checks on input data: </span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Columns: CHR, POS, RSID, EffectAllele, OtherAllele, B, SE, EAF, INFO, P, CaseN, and ControlN</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Examining data for missing values across all columns </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   • Using tidyr::drop_na()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no rows with missing values</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking for insertions/deletions ('indels') using:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   1. EffectAllele or OtherAllele, character length &gt; 1: A vs AA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   2. EffectAllele or OtherAllele coded as 'D', 'I', or 'R'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Detected 0 rows as indels</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Updating RSIDs that have been merged using RsMergeArch </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Updated RSID in 175 rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Finished initial checks</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   • 175 rows with updated RSID:</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">/tmp/RtmpdpV3FK/2023_08_01_18_13_36.612778/2023_08_01_18_13_36.612778/pipeline_info/updated_rsid.log.gz</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Starting validation of CHR,POS, RSID, EffectAllele and OtherAllele</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the RSID column: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; • Checking that RSID follows rs format: [rR][sS][1-10]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; • If rows fail rs format, look for CHR:POS or CHR:POS:REF:ALT format</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Found 307 rows with invalid RSID format: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Attempting to parse format...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Succeded in parsing 307 out of 307 rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Parsed format: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Found CHR and POS in those rows where parsing of RSID failing. Will use CHR and POS columns</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating CHR </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. CHR coerced to character `base::as.character()`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. converted to uppercase `stringr::str_to_upper()`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. 'CHR' removed (handles UCSC style CHR format)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 4. Check for values outside 1:22, X, Y</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 5. Check for NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass chr validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating POS </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. POS coerced to integer `base::as.integer()`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. Checks for NA, NaN, Inf</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. Checks for POS &lt;= 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 4. Check for POS &gt;= 10^9</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass pos validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating EffectAllele and OtherAllele: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. Will error if not type = character</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. EffectAllele and OtherAllele converted to uppercase</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. Check for EA/OA values that are not A,C,G or T</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 4. Check for NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass EA/OA validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Listing how many rows are removed per flag: </span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; invalid_rsid: 307</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; invalid_chr: 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; invalid_pos: 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; invalid_ea_oa: 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Removed a total of 307 rows:</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">/tmp/RtmpdpV3FK/2023_08_01_18_13_36.612778/2023_08_01_18_13_36.612778/pipeline_info/validate_snps.log.gz</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Validating statistics columns: B, P, SE, INFO, N, CaseN, ControlN, and EAF</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the EAF column: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. Will coerce to double: base::as.double()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. Check for EAF &lt;= 0 &amp; EAF &gt;= 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. Checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass EAF validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the SE column: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. Will coerce to double: base::as.double()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. Check for SE &lt;= 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. Checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass SE validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the P column: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. Coerce to double `base::as.double()`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. If P == 0, P is converted to `2.225074e-308`. Adjust by passing convert_p =</span></span>
<span><span class="co">#&gt; 'your_preferred_value'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 4. checks that P is within the range: P &gt;= 0 &amp; P &gt; 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass P validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the B column: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. Will coerce to double: base::as.double()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. Check for mislabelled OR by looking at median value</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 4. Print median value</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The median value of B is -0.00120072057651888, which seems reasonable</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass B validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the N column: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. Will coerce to integer: base::as.integer()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. Check for N &lt;= 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. Checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass N validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Listing how many rows are removed per flag: </span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> No rows were removed by these flags</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Finished tidyGWAS</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A total of 307 rows were removed. Started with: 100000 rows, ended with: 99693 rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Listing final breakdown of removed rows:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; validate_snps: 307</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre>
<p>If you dont provide a name for the sumstat cleaning, tidyGWAS uses
<code><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time()</a></code> to create a random name.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">my_outdir</span>, recursive <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "2023_08_01_18_13_36.612778/cleaned_GRCh38.gz"                  </span></span>
<span><span class="co">#&gt; [2] "2023_08_01_18_13_36.612778/pipeline_info/start_ids.tsv.gz"     </span></span>
<span><span class="co">#&gt; [3] "2023_08_01_18_13_36.612778/pipeline_info/updated_rsid.log.gz"  </span></span>
<span><span class="co">#&gt; [4] "2023_08_01_18_13_36.612778/pipeline_info/validate_snps.log.gz" </span></span>
<span><span class="co">#&gt; [5] "2023_08_01_18_13_36.612778/pipeline_info/validate_stats.log.gz"</span></span>
<span><span class="co">#&gt; [6] "2023_08_01_18_13_36.612778/raw_sumstats.gz"                    </span></span>
<span><span class="co">#&gt; [7] "2023_08_01_18_13_36.612778/tidyGWAS_logfile.txt"</span></span></code></pre></div>
<p>This can be controlled with the <code>name</code> argument:</p>
<pre><code><span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Output is redirected to logfile: <span style="color: #0000BB;">/tmp/RtmpdpV3FK/first_gwas/first_gwas/tidyGWAS_logfile.txt</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Running </span><span style="color: #0000BB; font-weight: bold;">tidyGWAS 0.1.0</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; Starting at 2023-08-01 18:13:39.655009</span></span>
<span><span class="co">#&gt; Writing out raw sumstats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Performing initial checks on input data: </span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Columns: CHR, POS, RSID, EffectAllele, OtherAllele, B, SE, EAF, INFO, P, CaseN, and ControlN</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Examining data for missing values across all columns </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   • Using tidyr::drop_na()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no rows with missing values</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Checking for insertions/deletions ('indels') using:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   1. EffectAllele or OtherAllele, character length &gt; 1: A vs AA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   2. EffectAllele or OtherAllele coded as 'D', 'I', or 'R'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Detected 0 rows as indels</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Updating RSIDs that have been merged using RsMergeArch </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Updated RSID in 175 rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Finished initial checks</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   • 175 rows with updated RSID:</span></span>
<span><span class="co">#&gt;   <span style="color: #0000BB;">/tmp/RtmpdpV3FK/first_gwas/first_gwas/pipeline_info/updated_rsid.log.gz</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Starting validation of CHR,POS, RSID, EffectAllele and OtherAllele</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the RSID column: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; • Checking that RSID follows rs format: [rR][sS][1-10]</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; • If rows fail rs format, look for CHR:POS or CHR:POS:REF:ALT format</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Found 307 rows with invalid RSID format: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Attempting to parse format...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Succeded in parsing 307 out of 307 rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Parsed format: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Found CHR and POS in those rows where parsing of RSID failing. Will use CHR and POS columns</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating CHR </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. CHR coerced to character `base::as.character()`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. converted to uppercase `stringr::str_to_upper()`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. 'CHR' removed (handles UCSC style CHR format)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 4. Check for values outside 1:22, X, Y</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 5. Check for NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass chr validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating POS </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. POS coerced to integer `base::as.integer()`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. Checks for NA, NaN, Inf</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. Checks for POS &lt;= 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 4. Check for POS &gt;= 10^9</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass pos validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating EffectAllele and OtherAllele: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. Will error if not type = character</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. EffectAllele and OtherAllele converted to uppercase</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. Check for EA/OA values that are not A,C,G or T</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 4. Check for NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass EA/OA validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Listing how many rows are removed per flag: </span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; invalid_rsid: 307</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; invalid_chr: 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; invalid_pos: 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; invalid_ea_oa: 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Removed a total of 307 rows:</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">/tmp/RtmpdpV3FK/first_gwas/first_gwas/pipeline_info/validate_snps.log.gz</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Validating statistics columns: B, P, SE, INFO, N, CaseN, ControlN, and EAF</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the EAF column: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. Will coerce to double: base::as.double()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. Check for EAF &lt;= 0 &amp; EAF &gt;= 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. Checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass EAF validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the SE column: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. Will coerce to double: base::as.double()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. Check for SE &lt;= 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. Checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass SE validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the P column: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. Coerce to double `base::as.double()`</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. If P == 0, P is converted to `2.225074e-308`. Adjust by passing convert_p =</span></span>
<span><span class="co">#&gt; 'your_preferred_value'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 4. checks that P is within the range: P &gt;= 0 &amp; P &gt; 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass P validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the B column: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. Will coerce to double: base::as.double()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. Check for mislabelled OR by looking at median value</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 4. Print median value</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The median value of B is -0.00120072057651888, which seems reasonable</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass B validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Validating the N column: </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 1. Will coerce to integer: base::as.integer()</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 2. Check for N &lt;= 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3. Checks for NA, NaN or Inf</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> All rows pass N validation</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">Listing how many rows are removed per flag: </span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> No rows were removed by these flags</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Finished tidyGWAS</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A total of 307 rows were removed. Started with: 100000 rows, ended with: 99693 rows</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Listing final breakdown of removed rows:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; validate_snps: 307</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt;  [1] "2023_08_01_18_13_36.612778/cleaned_GRCh38.gz"                  </span></span>
<span><span class="co">#&gt;  [2] "2023_08_01_18_13_36.612778/pipeline_info/start_ids.tsv.gz"     </span></span>
<span><span class="co">#&gt;  [3] "2023_08_01_18_13_36.612778/pipeline_info/updated_rsid.log.gz"  </span></span>
<span><span class="co">#&gt;  [4] "2023_08_01_18_13_36.612778/pipeline_info/validate_snps.log.gz" </span></span>
<span><span class="co">#&gt;  [5] "2023_08_01_18_13_36.612778/pipeline_info/validate_stats.log.gz"</span></span>
<span><span class="co">#&gt;  [6] "2023_08_01_18_13_36.612778/raw_sumstats.gz"                    </span></span>
<span><span class="co">#&gt;  [7] "2023_08_01_18_13_36.612778/tidyGWAS_logfile.txt"               </span></span>
<span><span class="co">#&gt;  [8] "first_gwas/cleaned_GRCh38.gz"                                  </span></span>
<span><span class="co">#&gt;  [9] "first_gwas/pipeline_info/start_ids.tsv.gz"                     </span></span>
<span><span class="co">#&gt; [10] "first_gwas/pipeline_info/updated_rsid.log.gz"                  </span></span>
<span><span class="co">#&gt; [11] "first_gwas/pipeline_info/validate_snps.log.gz"                 </span></span>
<span><span class="co">#&gt; [12] "first_gwas/pipeline_info/validate_stats.log.gz"                </span></span>
<span><span class="co">#&gt; [13] "first_gwas/raw_sumstats.gz"                                    </span></span>
<span><span class="co">#&gt; [14] "first_gwas/tidyGWAS_logfile.txt"</span></span></code></pre>
<p>Oops! We now have two different folders in our outdir. This is a
feature. <code>outdir</code> is meant to be used when you have many
different sumstats that you all want to be in the same directory, with
the same folder structure.</p>
<div class="section level3">
<h3 id="the-functions-within">The functions within<a class="anchor" aria-label="anchor" href="#the-functions-within"></a>
</h3>
<p>tidyGWAS exposes several smaller functions that can help when
interactively working with sumstats.</p>
</div>
<div class="section level3">
<h3 id="getting-b-and-se-from-z-n-and-eaf">Getting B and SE from Z, N and EAF<a class="anchor" aria-label="anchor" href="#getting-b-and-se-from-z-n-and-eaf"></a>
</h3>
<p><code><a href="../reference/repair_stats.html">repair_stats()</a></code> can be used to add any missing statistics
column, such as P, Z, B or SE, depending on which columns exist. Note
that there some</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Has Z , EAF and N, but not B and SE (Which are commonly used for PRS)</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">test_file</span>, Z <span class="op">=</span> <span class="va">B</span><span class="op">/</span><span class="va">SE</span>, N <span class="op">=</span> <span class="va">CaseN</span> <span class="op">+</span> <span class="va">ControlN</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">B</span>, <span class="op">-</span><span class="va">SE</span><span class="op">)</span></span>
<span><span class="va">repaired</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/repair_stats.html">repair_stats</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Repairing missing statistics columns:</span></span>
<span><span class="co">#&gt; 1. Transform OR to B if OR exists</span></span>
<span><span class="co">#&gt; 2. Remove OR if both OR and B exists</span></span>
<span><span class="co">#&gt; 3. Impute Z based on B and SE if both B and SE exist and Z is missing</span></span>
<span><span class="co">#&gt; 4. Impute Z based on P and B if Z and SE is missing</span></span>
<span><span class="co">#&gt; 5. Impute B and SE if both are missing, and Z, EAF and N is present</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Starting reparations:</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> <span style="font-weight: bold;">Imputing B and SE, using Z, EAF and N </span>. The B and SE will correspond to a standardized scale, which might not always be the same scale that the original betas was on.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Checking for NaN, Inf or NA in numeric columns</span></span>
<span><span class="co">#&gt; → Found no columns containing NA/Inf/NaN</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Finished repair_stats:</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Added 2 new columns: B and SE</span></span>
<span></span>
<span><span class="co"># the Betas will be on a standardized scale, and can therefore be different</span></span>
<span><span class="co"># than what was initially there</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="va">test_file</span>, <span class="va">repaired</span>,by <span class="op">=</span> <span class="st">"RSID"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">B.x</span>, <span class="va">B.y</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">B.x</span>, <span class="va">B.y</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cor(B.x, B.y)</span></span>
<span><span class="co">#&gt; 1     0.9956026</span></span>
<span></span>
<span></span>
<span><span class="co"># Getting Z, if you only have B and P</span></span></code></pre></div>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tmp</span> <span class="op">&lt;-</span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">test_file</span>,  <span class="op">-</span><span class="va">SE</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="../reference/repair_stats.html">repair_stats</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Repairing missing statistics columns:</span></span>
<span><span class="co">#&gt; 1. Transform OR to B if OR exists</span></span>
<span><span class="co">#&gt; 2. Remove OR if both OR and B exists</span></span>
<span><span class="co">#&gt; 3. Impute Z based on B and SE if both B and SE exist and Z is missing</span></span>
<span><span class="co">#&gt; 4. Impute Z based on P and B if Z and SE is missing</span></span>
<span><span class="co">#&gt; 5. Impute B and SE if both are missing, and Z, EAF and N is present</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Starting reparations:</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Found B and P but not Z. Imputing Z using:</span></span>
<span><span class="co">#&gt; sign(beta) * sqrt(stats::qchisq(pvalue,1,lower=FALSE))</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Checking for NaN, Inf or NA in numeric columns</span></span>
<span><span class="co">#&gt; → Found no columns containing NA/Inf/NaN</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Finished repair_stats:</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Added 1 new columns: Z</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="identifying-incorrect-rsids">Identifying incorrect RSIDs<a class="anchor" aria-label="anchor" href="#identifying-incorrect-rsids"></a>
</h3>
<p>uses a regex to identify any non RrSs rows.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">flag_incorrect_rsid_format</span><span class="op">(</span><span class="va">test_file</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">invalid_rsid</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="a-fast-function-for-finding-duplications-in-rsid-or-chr">A fast function for finding duplications in RSID or CHR<a class="anchor" aria-label="anchor" href="#a-fast-function-for-finding-duplications-in-rsid-or-chr"></a>
</h3>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/flag_duplicates.html">flag_duplicates</a></span><span class="op">(</span><span class="va">test_file</span>, column <span class="op">=</span> <span class="st">"rsid"</span><span class="op">)</span></span>
<span><span class="co">#&gt;         CHR       POS        RSID EffectAllele OtherAllele            B     SE</span></span>
<span><span class="co">#&gt;      1:   6  31819813 rs183988210            C           G -0.005696193 0.0292</span></span>
<span><span class="co">#&gt;      2:  16  13991338   rs8044430            T           C -0.017695649 0.0092</span></span>
<span><span class="co">#&gt;      3:   9  85016930 rs111404043            A           T -0.002302649 0.0185</span></span>
<span><span class="co">#&gt;      4:  19  15743403   rs4807961            C           G  0.002397125 0.0099</span></span>
<span><span class="co">#&gt;      5:  18   5948313    rs948293            T           C  0.001698557 0.0092</span></span>
<span><span class="co">#&gt;     ---                                                                       </span></span>
<span><span class="co">#&gt;  99996:  10 132193006   rs7904068            T           C  0.020096702 0.0102</span></span>
<span><span class="co">#&gt;  99997:  17  31324100  rs16967856            C           T  0.006101349 0.0252</span></span>
<span><span class="co">#&gt;  99998:   7  33310046  rs10246533            G           A -0.008102739 0.0086</span></span>
<span><span class="co">#&gt;  99999:  11 107469336    rs597383            G           C -0.005404578 0.0159</span></span>
<span><span class="co">#&gt; 100000:   2 101246512  rs10173962            C           T  0.004798469 0.0090</span></span>
<span><span class="co">#&gt;               EAF  INFO       P CaseN ControlN dup_rsid</span></span>
<span><span class="co">#&gt;      1: 0.9754088 0.936 0.84440 51235    74087    FALSE</span></span>
<span><span class="co">#&gt;      2: 0.6407741 0.951 0.05315 53386    77258    FALSE</span></span>
<span><span class="co">#&gt;      3: 0.9160000 0.701 0.89970 53386    77258    FALSE</span></span>
<span><span class="co">#&gt;      4: 0.2556345 0.996 0.80740 53386    77258    FALSE</span></span>
<span><span class="co">#&gt;      5: 0.3318173 0.971 0.85780 53386    77258    FALSE</span></span>
<span><span class="co">#&gt;     ---                                                </span></span>
<span><span class="co">#&gt;  99996: 0.7658173 0.991 0.04839 53386    77258    FALSE</span></span>
<span><span class="co">#&gt;  99997: 0.9670000 0.930 0.80840 53386    77258    FALSE</span></span>
<span><span class="co">#&gt;  99998: 0.4740000 0.999 0.34810 53386    77258    FALSE</span></span>
<span><span class="co">#&gt;  99999: 0.9201827 0.997 0.73210 53386    77258    FALSE</span></span>
<span><span class="co">#&gt; 100000: 0.6391827 0.995 0.58960 53386    77258    FALSE</span></span>
<span><span class="co"># or </span></span>
<span><span class="fu"><a href="../reference/flag_duplicates.html">flag_duplicates</a></span><span class="op">(</span><span class="va">test_file</span>, column <span class="op">=</span> <span class="st">"chr_pos"</span><span class="op">)</span></span>
<span><span class="co">#&gt;         CHR       POS        RSID EffectAllele OtherAllele            B     SE</span></span>
<span><span class="co">#&gt;      1:   6  31819813 rs183988210            C           G -0.005696193 0.0292</span></span>
<span><span class="co">#&gt;      2:  16  13991338   rs8044430            T           C -0.017695649 0.0092</span></span>
<span><span class="co">#&gt;      3:   9  85016930 rs111404043            A           T -0.002302649 0.0185</span></span>
<span><span class="co">#&gt;      4:  19  15743403   rs4807961            C           G  0.002397125 0.0099</span></span>
<span><span class="co">#&gt;      5:  18   5948313    rs948293            T           C  0.001698557 0.0092</span></span>
<span><span class="co">#&gt;     ---                                                                       </span></span>
<span><span class="co">#&gt;  99996:  10 132193006   rs7904068            T           C  0.020096702 0.0102</span></span>
<span><span class="co">#&gt;  99997:  17  31324100  rs16967856            C           T  0.006101349 0.0252</span></span>
<span><span class="co">#&gt;  99998:   7  33310046  rs10246533            G           A -0.008102739 0.0086</span></span>
<span><span class="co">#&gt;  99999:  11 107469336    rs597383            G           C -0.005404578 0.0159</span></span>
<span><span class="co">#&gt; 100000:   2 101246512  rs10173962            C           T  0.004798469 0.0090</span></span>
<span><span class="co">#&gt;               EAF  INFO       P CaseN ControlN dup_chr_pos</span></span>
<span><span class="co">#&gt;      1: 0.9754088 0.936 0.84440 51235    74087       FALSE</span></span>
<span><span class="co">#&gt;      2: 0.6407741 0.951 0.05315 53386    77258       FALSE</span></span>
<span><span class="co">#&gt;      3: 0.9160000 0.701 0.89970 53386    77258       FALSE</span></span>
<span><span class="co">#&gt;      4: 0.2556345 0.996 0.80740 53386    77258       FALSE</span></span>
<span><span class="co">#&gt;      5: 0.3318173 0.971 0.85780 53386    77258       FALSE</span></span>
<span><span class="co">#&gt;     ---                                                   </span></span>
<span><span class="co">#&gt;  99996: 0.7658173 0.991 0.04839 53386    77258       FALSE</span></span>
<span><span class="co">#&gt;  99997: 0.9670000 0.930 0.80840 53386    77258       FALSE</span></span>
<span><span class="co">#&gt;  99998: 0.4740000 0.999 0.34810 53386    77258       FALSE</span></span>
<span><span class="co">#&gt;  99999: 0.9201827 0.997 0.73210 53386    77258       FALSE</span></span>
<span><span class="co">#&gt; 100000: 0.6391827 0.995 0.58960 53386    77258       FALSE</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Arvid Harder.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
