<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="tidyGWAS">
<title>tidyGWAS • tidyGWAS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="tidyGWAS">
<meta property="og:description" content="tidyGWAS">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tidyGWAS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/tidyGWAS.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/transforming_dbsnp_to_parquet.html">get-to-know-dbsnp</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>tidyGWAS</h1>
            
      
      
      <div class="d-none name"><code>tidyGWAS.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>tidyGWAS dependencies should be easily installed on most systems,
with the arrow dependecy being the most likely to cause trouble. In the
case that installation of arrow is causing trouble, see <a href="https://arrow.apache.org/docs/r/articles/install.html" class="external-link">here</a>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ararder/tidyGWAS"</span><span class="op">)</span></span>
<span><span class="co"># or</span></span>
<span><span class="co"># install.packages("remotes")</span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"ararder/tidyGWAS"</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="dbsnp155-reference-file">dbSNP155 reference file<a class="anchor" aria-label="anchor" href="#dbsnp155-reference-file"></a>
</h4>
<p>tidyGWAS uses a version dbSNP 155 converted to the <a href="https://arrow.apache.org" class="external-link">Apache Arrow</a> .parquet files. You can
download the dbSNP155 reference file from inside R using the
<code>googledrive</code> package, or by manually navigating to this <a href="https://drive.google.com/file/d/1LmgfmTQaWwJaFpBHcRQIY_kwe5iN7Pj6/view?usp=share_link" class="external-link">file</a></p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># You can download the file from inside R using the googledrive package:</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://googledrive.tidyverse.org" class="external-link">googledrive</a></span><span class="op">)</span></span>
<span><span class="fu">googledrive</span><span class="fu">::</span><span class="fu">drive_deauth</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">id</span> <span class="op">&lt;-</span> <span class="fu">googledrive</span><span class="fu">::</span><span class="fu">as_id</span><span class="op">(</span><span class="st">"1LmgfmTQaWwJaFpBHcRQIY_kwe5iN7Pj6"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##### EDIT THIS:</span></span>
<span><span class="va">filepath_to_store_dir</span> <span class="op">&lt;-</span> <span class="st">""</span></span>
<span><span class="co">##### ---------------------</span></span>
<span></span>
<span><span class="fu">drive_download</span><span class="op">(</span><span class="va">id</span>, path <span class="op">=</span> <span class="va">filepath_to_store_dir</span><span class="op">)</span></span></code></pre></div>
<p>The file needs to be untarred</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># change directory to where the downloaded file is</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tar</span> <span class="at">-xf</span> dbSNP155.tar dbSNP155</span></code></pre></div>
<p>You should provide the full path to the directory
<code>dbSNP155</code>, which should contain the following folders after
untarring.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org" class="external-link">fs</a></span><span class="op">)</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># this is how the untarred folder should look if you ls inside it</span></span>
<span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_ls.html" class="external-link">dir_ls</a></span><span class="op">(</span><span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"tidyGWAS"</span><span class="op">)</span>, <span class="st">"extdata/dbSNP155"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; /home/runner/work/_temp/Library/tidyGWAS/extdata/dbSNP155/GRCh37</span></span>
<span><span class="co">#&gt; /home/runner/work/_temp/Library/tidyGWAS/extdata/dbSNP155/GRCh38</span></span>
<span><span class="co">#&gt; /home/runner/work/_temp/Library/tidyGWAS/extdata/dbSNP155/refsnp-merged</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://arvidharder.com/tidyGWAS/">tidyGWAS</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://fs.r-lib.org" class="external-link">fs</a></span><span class="op">)</span></span>
<span><span class="va">example_file</span> <span class="op">&lt;-</span> <span class="fu">tidyGWAS</span><span class="fu">::</span><span class="va"><a href="../reference/test_file.html">test_file</a></span></span>
<span></span>
<span><span class="co"># Change this to fit your own system. Here a very small demo version is used</span></span>
<span><span class="va">dbsnp_file</span> <span class="op">&lt;-</span>  <span class="fu"><a href="https://fs.r-lib.org/reference/path.html" class="external-link">path</a></span><span class="op">(</span><span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/path_package.html" class="external-link">path_package</a></span><span class="op">(</span><span class="st">"tidyGWAS"</span><span class="op">)</span>, <span class="st">"extdata/dbSNP155"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span>
<span>  <span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  <span class="co"># only tbl (the data.frame), is a mandatory argument.</span></span>
<span>  tbl <span class="op">=</span> <span class="va">example_file</span>, </span>
<span>  <span class="co"># However, if the dbsnp_path argument is not passed,</span></span>
<span>  <span class="co"># no comparisons against dbSNP155 will be done, or any</span></span>
<span>  <span class="co"># merged RSIDs will be updated</span></span>
<span>  dbsnp_path <span class="op">=</span> <span class="va">dbsnp_file</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="a-more-typical-functional-call">A more typical functional call<a class="anchor" aria-label="anchor" href="#a-more-typical-functional-call"></a>
</h4>
<p>Typically, you will call <code><a href="../reference/tidyGWAS.html">tidyGWAS()</a></code> with a few more
arguments. By default <code><a href="../reference/tidyGWAS.html">tidyGWAS()</a></code> will use
<code><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">base::tempdir()</a></code> to write to during execution. In addition,
the cleaned sumstat is also returned by the function. The final output
will also be saved in this tempdir. Since its a temporary directory, you
will loose all files if you exit the R Session. This behaviour can be
adjusteded with the <code>outdir</code> argument.</p>
<p>Let’s take a look at how a more typical call to tidyGWAS would
look:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># here is where you would substite your own sumstats, either as a filepath,</span></span>
<span><span class="co"># or in-memory data.frame</span></span>
<span></span>
<span><span class="co"># your own sumstat</span></span>
<span><span class="va">your_sumstats</span> <span class="op">&lt;-</span> <span class="fu">tidyGWAS</span><span class="fu">::</span><span class="va"><a href="../reference/test_file.html">test_file</a></span></span>
<span></span>
<span><span class="co"># a directory where you want your data to be stored after succesful cleaning.</span></span>
<span><span class="va">gwas_folder</span> <span class="op">&lt;-</span> <span class="fu">withr</span><span class="fu">::</span><span class="fu"><a href="https://withr.r-lib.org/reference/with_tempfile.html" class="external-link">local_tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">cleaned</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">your_sumstats</span>, </span>
<span>  dbsnp_path <span class="op">=</span> <span class="va">dbsnp_file</span>,</span>
<span>  <span class="co"># it's useful to write all the messages to a logfile, so messages are saved.</span></span>
<span>  logfile<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>  <span class="co"># by default, the name of the output folder is a concotonated call to Sys.time()</span></span>
<span>  <span class="co"># but you can adjust this by using the name argument</span></span>
<span>  name <span class="op">=</span> <span class="st">"my_first_tidyGWAS"</span>,</span>
<span>  <span class="co"># by default, the saved file is written out as a gzipped csv file,</span></span>
<span>  <span class="co"># however, the arrow parquet format offers many advantages over the classical csv file</span></span>
<span>  output_format <span class="op">=</span> <span class="st">"parquet"</span>,</span>
<span>  <span class="co"># by passing the argument outdir, you can tell tidyGWAS to copy over files</span></span>
<span>  <span class="co"># after a finished execution to that directory</span></span>
<span>  outdir <span class="op">=</span> <span class="va">gwas_folder</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Output is redirected to logfile: <span style="color: #0000BB;">/tmp/RtmpCDpzK0/my_first_tidyGWAS/my_first_tidyGWAS/tidyGWAS_logfile.txt</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Running </span><span style="color: #0000BB; font-weight: bold;">tidyGWAS 0.3.0</span> <span style="color: #00BBBB;">──────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; Starting at 2023-09-27 05:43:19.92775, with 100000 rows in input data.frame</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Saving all files during execution to <span style="color: #0000BB;">/tmp/RtmpCDpzK0/my_first_tidyGWAS/my_first_tidyGWAS</span>,</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> After execution, files will be copied to <span style="color: #0000BB;">/tmp/RtmpCDpzK0/file19c3129b9ccc/my_first_tidyGWAS</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 1) Checking that columns follow tidyGWAS format </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> The following columns are used for further steps: <span style="font-style: italic;">CHR</span>, <span style="font-style: italic;">POS</span>, <span style="font-style: italic;">RSID</span>, <span style="font-style: italic;">EffectAllele</span>, <span style="font-style: italic;">OtherAllele</span>, <span style="font-style: italic;">B</span>, <span style="font-style: italic;">SE</span>, <span style="font-style: italic;">EAF</span>, <span style="font-style: italic;">INFO</span>, <span style="font-style: italic;">P</span>, <span style="font-style: italic;">CaseN</span>, <span style="font-style: italic;">ControlN</span>, and <span style="font-style: italic;">rowid</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">1) Scanning for rows with NA</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> No rows contained missing values</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">2) Updating merged RSIDs</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> 175 rows with updated RSID</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; •</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">/tmp/RtmpCDpzK0/my_first_tidyGWAS/my_first_tidyGWAS/pipeline_info/updated_rsid.parquet</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">3) Scanning for rows with duplications</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A unique ID is formed by concontenating CHR, POS, EffectAllele, and OtherAllele</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Found no duplications</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">4) Scanning for indels</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   1. EffectAllele or OtherAllele, character length &gt; 1: A vs AA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   2. EffectAllele or OtherAllele coded as 'D', 'I', or 'R'</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Detected 0 rows as indels</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">5) Scanning for invalid RSID </span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> 307 rows had an invalid RSID. RSID will be repaired using dbSNP if possible</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">6) Column validation is done separately for main rows, rows without RSID and indels</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 6) Validating columns: for main rows </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The median value of B is -0.0012, which seems reasonable</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-style: italic;">All rows passed validation</span> </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 6) Validating columns: for rows without RSID </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> The median value of B is -0.0022, which seems reasonable</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-style: italic;">All rows passed validation</span> </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">7 Using dbSNP to repair and validate CHR:POS:RSID</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; • Variants with CHR and POS or RSID not present in dbSNP are removed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; • Checking that EffectAllele and OtherAllele is compatible with REF and ALT in</span></span>
<span><span class="co">#&gt; dbSNP</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 7a) Starting with main rows:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Using CHR, POS and RSID to align with dbSNP</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Inferring build by matching 10000 rows to GRCh37 and GRCh38</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 3363 snps matched GRCh38, 9999 for GRCh37, inferring build to be 37</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A total of 11 rows had incompatible CHR:POS and RSID data. These have been updated</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Listing how many rows are removed per flag:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   no_dbsnp_entry: 11</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; • Removed a total of 11 rows:</span></span>
<span><span class="co">#&gt; <span style="color: #0000BB;">/tmp/RtmpCDpzK0/my_first_tidyGWAS/my_first_tidyGWAS/pipeline_info/removed_rows_main_validate_with_dbsnp.parquet</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── 7b) rows without RSID:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Using CHR and POS to align with dbSNP</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Inferring build by matching 10000 rows to GRCh37 and GRCh38</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; 101 snps matched GRCh38, 306 for GRCh37, inferring build to be 37</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-style: italic;">All rows passed validation</span> </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── <span style="font-weight: bold;">8) Repairing missings columns if possible</span> ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Z missing: Calculating Z using the formula:  Z = B / SE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Finished repair_stats:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Added 1 new columns: Z</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">✔</span> Detected 0 rows as indels</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Listing final breakdown of removed rows:  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   main_validate_with_dbsnp: 11</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Finished tidyGWAS</span> <span style="color: #00BBBB;">───────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> A total of 11 rows were removed</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">ℹ</span> Total running time: 10.6s</span></span></code></pre></div>
<p>Now we have saved our output in <code>gwas_folder</code> . By using
the <code>name</code> and <code>outdir</code> argument, tidyGWAS has
saved all the output in our outdir, in a directory named
“my_first_tidyGWAS”.</p>
<p>Inside this folder we have several files that can be used to
understand what tidyGWAS has done.</p>
<p>tidyGWAS_logfile.txt contains the messages you see printed above.</p>
<p>in /pipeline_info/ we can find files containing the rowid of all rows
that were removed.</p>
<p>In /raw_sumstats.parquet, we can find the raw sumstats, as we passed
it to <code><a href="../reference/tidyGWAS.html">tidyGWAS()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://fs.r-lib.org/reference/dir_ls.html" class="external-link">dir_ls</a></span><span class="op">(</span><span class="va">gwas_folder</span>, recurse <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; /tmp/RtmpCDpzK0/file19c3129b9ccc/my_first_tidyGWAS</span></span>
<span><span class="co">#&gt; /tmp/RtmpCDpzK0/file19c3129b9ccc/my_first_tidyGWAS/cleaned_GRCh38.parquet</span></span>
<span><span class="co">#&gt; /tmp/RtmpCDpzK0/file19c3129b9ccc/my_first_tidyGWAS/pipeline_info</span></span>
<span><span class="co">#&gt; /tmp/RtmpCDpzK0/file19c3129b9ccc/my_first_tidyGWAS/pipeline_info/removed_rows_main_col_validation.parquet</span></span>
<span><span class="co">#&gt; /tmp/RtmpCDpzK0/file19c3129b9ccc/my_first_tidyGWAS/pipeline_info/removed_rows_main_validate_with_dbsnp.parquet</span></span>
<span><span class="co">#&gt; /tmp/RtmpCDpzK0/file19c3129b9ccc/my_first_tidyGWAS/pipeline_info/removed_rows_without_rsid_col_validation.parquet</span></span>
<span><span class="co">#&gt; /tmp/RtmpCDpzK0/file19c3129b9ccc/my_first_tidyGWAS/pipeline_info/removed_rows_without_rsid_validate_with_dbsnp.parquet</span></span>
<span><span class="co">#&gt; /tmp/RtmpCDpzK0/file19c3129b9ccc/my_first_tidyGWAS/pipeline_info/updated_rsid.parquet</span></span>
<span><span class="co">#&gt; /tmp/RtmpCDpzK0/file19c3129b9ccc/my_first_tidyGWAS/raw_sumstats.parquet</span></span>
<span><span class="co">#&gt; /tmp/RtmpCDpzK0/file19c3129b9ccc/my_first_tidyGWAS/tidyGWAS_logfile.txt</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="creating-the-correct-column-names">Creating the correct column names<a class="anchor" aria-label="anchor" href="#creating-the-correct-column-names"></a>
</h3>
<p>tidyGWAS does not automatically detect column names, but expect the
<code><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">dplyr::tibble()</a></code> to contain the correct column names.
<code>tidyGWAS_columns</code> is a helper function created to match
column name to the tidyGWAS format.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># what if we have a file where POS was named bp, and B was named BETA?</span></span>
<span><span class="co"># tidyGWAS will not recognize these as valid column names and would drop them.</span></span>
<span><span class="va">wrong_colnames</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span><span class="va">example_file</span>, bp <span class="op">=</span> <span class="va">POS</span>, BETA <span class="op">=</span> <span class="va">B</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">wrong_colnames</span><span class="op">)</span></span>
<span><span class="co">#&gt;   CHR       bp        RSID EffectAllele OtherAllele         BETA     SE</span></span>
<span><span class="co">#&gt; 1   6 31819813 rs183988210            C           G -0.005696193 0.0292</span></span>
<span><span class="co">#&gt; 2  16 13991338   rs8044430            T           C -0.017695649 0.0092</span></span>
<span><span class="co">#&gt; 3   9 85016930 rs111404043            A           T -0.002302649 0.0185</span></span>
<span><span class="co">#&gt; 4  19 15743403   rs4807961            C           G  0.002397125 0.0099</span></span>
<span><span class="co">#&gt; 5  18  5948313    rs948293            T           C  0.001698557 0.0092</span></span>
<span><span class="co">#&gt; 6  13 47224772       rs837            G           A  0.027196792 0.0086</span></span>
<span><span class="co">#&gt;         EAF  INFO        P CaseN ControlN</span></span>
<span><span class="co">#&gt; 1 0.9754088 0.936 0.844400 51235    74087</span></span>
<span><span class="co">#&gt; 2 0.6407741 0.951 0.053150 53386    77258</span></span>
<span><span class="co">#&gt; 3 0.9160000 0.701 0.899700 53386    77258</span></span>
<span><span class="co">#&gt; 4 0.2556345 0.996 0.807400 53386    77258</span></span>
<span><span class="co">#&gt; 5 0.3318173 0.971 0.857800 53386    77258</span></span>
<span><span class="co">#&gt; 6 0.4470432 1.010 0.001594 53386    77258</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we can use tidyGWAS_columns to specify current names:</span></span>
<span><span class="va">correct_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/tidyGWAS_columns.html">tidyGWAS_columns</a></span><span class="op">(</span></span>
<span>  <span class="co"># first argument is the data.frame</span></span>
<span>  tbl <span class="op">=</span> <span class="va">wrong_colnames</span>, </span>
<span>  <span class="co"># then any column which is not correctly named:</span></span>
<span>  POS <span class="op">=</span> <span class="st">"bp"</span>,</span>
<span>  B <span class="op">=</span> <span class="st">"BETA"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">correct_names</span><span class="op">)</span></span>
<span><span class="co">#&gt;   CHR      POS        RSID EffectAllele OtherAllele            B     SE</span></span>
<span><span class="co">#&gt; 1   6 31819813 rs183988210            C           G -0.005696193 0.0292</span></span>
<span><span class="co">#&gt; 2  16 13991338   rs8044430            T           C -0.017695649 0.0092</span></span>
<span><span class="co">#&gt; 3   9 85016930 rs111404043            A           T -0.002302649 0.0185</span></span>
<span><span class="co">#&gt; 4  19 15743403   rs4807961            C           G  0.002397125 0.0099</span></span>
<span><span class="co">#&gt; 5  18  5948313    rs948293            T           C  0.001698557 0.0092</span></span>
<span><span class="co">#&gt; 6  13 47224772       rs837            G           A  0.027196792 0.0086</span></span>
<span><span class="co">#&gt;          P       EAF CaseN ControlN  INFO</span></span>
<span><span class="co">#&gt; 1 0.844400 0.9754088 51235    74087 0.936</span></span>
<span><span class="co">#&gt; 2 0.053150 0.6407741 53386    77258 0.951</span></span>
<span><span class="co">#&gt; 3 0.899700 0.9160000 53386    77258 0.701</span></span>
<span><span class="co">#&gt; 4 0.807400 0.2556345 53386    77258 0.996</span></span>
<span><span class="co">#&gt; 5 0.857800 0.3318173 53386    77258 0.971</span></span>
<span><span class="co">#&gt; 6 0.001594 0.4470432 53386    77258 1.010</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="parallel-computation">Parallel computation<a class="anchor" aria-label="anchor" href="#parallel-computation"></a>
</h2>
<p>The Apache Arrow C++ implementation, automatically detects and uses
multiple cores. In some cases, for example when running tidyGWAS in a
HPC cluster, you might need to manually set the number of cores, which
can be done using the <code>OMP_NUM_THREADS</code> variable. This should
not be a larger number of cores than what you have requested in your HPC
job (in the example below, the “–cpus-per-task” flag)</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --mem=60gb</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --time=24:0:00</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#SBATCH --cpus-per-task 8</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">OMP_NUM_THREADS</span><span class="op">=</span>8</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="va">outdir</span><span class="op">=</span><span class="va">$(</span><span class="bu">pwd</span><span class="va">)</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="va">gwas</span><span class="op">=</span><span class="va">$outdir</span>/my_gwas.tsv.gz</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="va">dbsnp_files</span><span class="op">=</span><span class="st">"dbSNP155"</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">"tidyGWAS(commandArgs(trailingOnly = TRUE)[1],  dbsnp_path = commandArgs(trailingOnly = TRUE)[2],outdir = commandArgs(trailingOnly = TRUE)[3], logfile=TRUE)"</span> <span class="va">$gwas</span> <span class="va">$dbsnp_files</span> <span class="va">$outdir</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="computational-cost">Computational cost<a class="anchor" aria-label="anchor" href="#computational-cost"></a>
</h2>
<p>Memory use and time scales with the size of the summary statistics.
From running tidyGWAS experimentally, here’s an estimation of time and
memory.</p>
<div class="section level5">
<h5 id="cpu-usage">CPU usage<a class="anchor" aria-label="anchor" href="#cpu-usage"></a>
</h5>
<p>CPU usage depends heavily on CPU type and number of available
cores.</p>
<ol style="list-style-type: decimal">
<li>Macbook Pro M2: 8 million rows, ~4.42 minutes, &lt; 16GB memory
used</li>
<li>AMD EPYC 75F3 32-Core Processor; ~ 6 minutes &lt; 16GB memory
used</li>
</ol>
<p>For 60 million rows:</p>
<ol style="list-style-type: decimal">
<li>AMD EPYC 75F3 32-Core Processor; ~ 34 minutes &lt; 75gb using 5
cores</li>
</ol>
</div>
<div class="section level4">
<h4 id="memory-usage-and-number-of-rows">Memory usage and number of rows<a class="anchor" aria-label="anchor" href="#memory-usage-and-number-of-rows"></a>
</h4>
<p>The memory usage will likely be the main constraint with increasing
size of summary statistics.</p>
<ol style="list-style-type: decimal">
<li>20 million rows &lt; 20gb</li>
<li>40 million rows &lt; 40gb</li>
<li>60 million rows &lt; 75gb</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="reading-in-files-from-disk">Reading in files from disk<a class="anchor" aria-label="anchor" href="#reading-in-files-from-disk"></a>
</h2>
<p>Behind the scenes, tidyGWAS uses
<code><a href="https://arrow.apache.org/docs/r/reference/read_delim_arrow.html" class="external-link">arrow::read_delim_arrow()</a></code> to read in the file using the
filepath provided.</p>
<p>Through <code>...</code>, you can pass specific arguments if your
file requires adjustments to read in, such as <code>delim = "\t"</code>.
You can read more about this <a href="https://arrow.apache.org/docs/r/reference/read_delim_arrow.html" class="external-link">here</a>.
In general, if your GWAS is in a file format that tidyGWAS struggles to
read in properly, it is often easier toread in the file using
<code>data.table::fread()</code> or <code>readr::read_tsv</code>, or any
package that focus specifically on reading in files, and then passing
the data.frame to tidyGWAS in-memory.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu">data.table</span><span class="fu">::</span><span class="fu">fread</span><span class="op">(</span><span class="st">"/file/that/is/challenging_to_read.vcf.gz"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/tidyGWAS.html">tidyGWAS</a></span><span class="op">(</span></span>
<span>  tbl <span class="op">=</span> <span class="va">file</span>,</span>
<span>  dbsnp_path <span class="op">=</span> <span class="va">dbsnp_files</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="useful-functions">Useful functions<a class="anchor" aria-label="anchor" href="#useful-functions"></a>
</h2>
<p>tidyGWAS exposes many of it’s internal functions, if there is only a
specific part you want to use.</p>
<div class="section level4">
<h4 id="getting-b-and-se-from-z-n-and-eaf">Getting B and SE from Z, N and EAF<a class="anchor" aria-label="anchor" href="#getting-b-and-se-from-z-n-and-eaf"></a>
</h4>
<p><code><a href="../reference/repair_stats.html">repair_stats()</a></code> can be used to add any missing statistics
column, such as P, Z, B or SE, depending on which columns exist. Note
that this imputation is not error free, and will not perfectly impute
missing data.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_file</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="va">test_file</span><span class="op">)</span></span>
<span><span class="co"># Has Z , EAF and N, but not B and SE (Which are commonly used for PRS)</span></span>
<span><span class="va">tmp</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span><span class="va">example_file</span>, Z <span class="op">=</span> <span class="va">B</span><span class="op">/</span><span class="va">SE</span>, N <span class="op">=</span> <span class="va">CaseN</span> <span class="op">+</span> <span class="va">ControlN</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">B</span>, <span class="op">-</span><span class="va">SE</span><span class="op">)</span></span>
<span><span class="va">repaired</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/repair_stats.html">repair_stats</a></span><span class="op">(</span><span class="va">tmp</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># the Betas will be on a standardized scale, and can therefore be different</span></span>
<span><span class="co"># than what was initially there</span></span>
<span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">inner_join</a></span><span class="op">(</span><span class="va">example_file</span>, <span class="va">repaired</span>,by <span class="op">=</span> <span class="st">"RSID"</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">B.x</span>, <span class="va">B.y</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">B.x</span>, <span class="va">B.y</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># Getting Z, if you only have B and P</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Arvid Harder.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
